---
title: "*A.californica* transcriptome analysis"
author: Javier A. Rodriguez Casariego
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
---

# Setup

```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/JARC_2T/NSF-CREST_Postdoc/LongTerm_Hypoxia_exp/AplCal_hypoxia/TagSeq/')
```

```{r load_libraries, include = TRUE}
# Load libraries
library(DESeq2)    # BiocManager::install('DESeq2')
library(limma)     # BiocManager::install('limma')
library(stringr)
library(readxl)
library(pheatmap)
library(RColorBrewer)
library(variancePartition)  # BiocManager::install('variancePartition')
library(doParallel)
library(cowplot)
source("analyses/GO_MWU/gomwu.functions.R")
library(KOGMWU)  # install.packages("KOGMWU")
library(adegenet)
library(ggpubr)
library(kableExtra)
library(tidyverse)
library(ggfortify) # install.packages("ggfortify")
library(Biobase)
library(arrayQualityMetrics)
library(vegan)
library(ggvenn) # install.packages("ggvenn")
if (!require(devtools)) install.packages("devtools")
devtools::install_github("gaospecial/ggVennDiagram")
library(ggVennDiagram)

## ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10) %+replace%    #, base_family = "Arial"
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA),
      axis.text.x = element_text(angle=45, hjust=1, vjust = 1)
    )
}
## ggplot labeller
colnames <- c(
  `A` = "abdominal",
  `Pp` = "pleural & pedal"
)

parents <- c(
  `60` = "lab-reared",
  `71` = "wild"
)

global_labeller <- labeller(
  tissue = colnames,
  batch = parents,
  .default = "label_parsed"
)
```

# Load data

#### Gene count data
```{r import_counts}
#name <- gsub( "(?:[^_]+_){4}([^_ ]+)*$","", files)
  
  # STAR quantMode geneCounts output:
  #column 1: gene ID
  #column 2: counts for unstranded RNA-seq
  #column 3: counts for the 1st read strand aligned with RNA (htseq-count option -s yes)
  #column 4: counts for the 2nd read strand aligned with RNA (htseq-count option -s reverse)
  
sampleNames <- list.files(path = glue::glue(getwd(), "/data/GeneCounts"), pattern = "*.ReadsPerGene.out.tab") %>%
    stringr::str_split_fixed("_", n = 4) %>%
    tibble::as_tibble() %>%
    dplyr::mutate(Name = V1) %>%
    dplyr::select(Name) %>% 
    purrr::flatten_chr()
  
geneIDs <- list.files(path = glue::glue(getwd(), "/data/GeneCounts"), pattern = "*.ReadsPerGene.out.tab", full.names = T)[1] %>% 
    data.table::fread(select = 1) %>%
    purrr::flatten_chr()
  
countMatrix <- list.files(path = glue::glue(getwd(), "/data/GeneCounts"), pattern = "*.ReadsPerGene.out.tab", full.names = T) %>%
    purrr::map_dfc(data.table::fread, select = 3, data.table = F) %>%
    magrittr::set_colnames(sampleNames) %>% 
    magrittr::set_rownames(geneIDs)
  
countMatrix <- countMatrix[-c(1:4),]
```

#### Sample metadata
```{r sample_metadata}
# Import sample metadata
sdat0 <- read_csv("data/sample_metadata.csv") 
sdat0$treatment <- gsub("C", "Control", sdat0$treatment)
sdat0 <- sdat0 %>%  
  mutate(tissue.group = interaction(tissue, treatment)) %>%
  mutate(batch.group = interaction(batch, treatment)) %>%
  mutate_if(is.character, as.factor)

sdat <- sdat0 %>%
  arrange(sample_ID) %>%                              # order rows by sample name
  column_to_rownames(var = "sample_ID")               # set sample name to rownames

save(sdat,sdat0,countMatrix, file = "output/counts_and_meta.RData")

#sdat$treatment <- factor(sdat$treatment, levels = c("Control","Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery"), ordered = T)
#sdat$group <- paste(sdat$treatment, sdat$batch, sdat$tissue, sep = "")
```

#### Gene annotations
```{r import_annot and modify}

annot <- read.delim("../../../Reference_Datasets/annotations_processed/AplCal3_gene2tx2prot2hs_uniprot.txt")
head(annot)

gene2uniprot <- annot[,c(1,5)]
#write_delim(gene2uniprot,"analyses/annotation/AplCal3_gene2uniprot.txt", delim = "\t")
#write(gene2uniprot$UNIPROT,"analyses/annotation/AplCal3_uniprot.txt")

uni2go <- read.delim("../../../Reference_Datasets/annotations_processed/uniprot-compressed_true_download_true_fields_accession_2Cid_2Cprotei-2022.10.01-15.56.44.60.tsv")
colnames(uni2go)[2] <- "UNIPROT"

prot2ko <- read.delim("../../../Reference_Datasets/annotations_processed/AplCal3.0_prot2ko.tab")
colnames(prot2ko)[1] <- "prot_version"

gene2go <- merge(gene2uniprot, uni2go, by = "UNIPROT", all.x = T)
head(gene2go)
gene2go <- gene2go[,c(2,8)]
gene2go$Gene.Ontology.IDs <- gsub(" ","", gene2go$Gene.Ontology.IDs)
gene2go <- gene2go[rowSums(is.na(gene2go)) == 0,]
#write_delim(gene2go,"analyses/GO_MWU/seq2go.tab", delim = "\t", col_names = F)

# build extended annotation

annot_full <- merge(annot, uni2go, by="UNIPROT", all.x=T)
annot_full <- annot_full[,c(2:8,1,11,13,14)]
annot_full <- merge(annot_full, prot2ko, by="prot_version", all.x = T)
annot_full <- annot_full[,c(2:6,1,7:12)] %>%
  distinct(gene, .keep_all = TRUE)

#write_delim(annot_full,"analyses/annotation/AplCal3_full_annotation.tsv", delim = "\t")

# get KOG annotation
prot2kog <- read.delim("../../../Reference_Datasets/AplCal3_JARC_gene2kogClass.tab", header = F)
colnames(prot2kog) <- c("prot_version","KOG_class")
gene2kog <- merge(annot_full, prot2kog, by = "prot_version", all.x=T)
gene2kog <- gene2kog[,c(2,13)]
write_delim(gene2kog,"analyses/annotation/AplCal3_gene2kog.tsv", delim = "\t")
```

### Create DESeqDataSet
```{r subset_dds}

# Create full DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countMatrix,
                              colData = sdat,
                              design = ~ tissue)
### Remove genes counted zero times
dds <- dds[ rowSums(counts(dds)) > 0, ]

# Subset DESeqDataSet
## Subset for batch and ganglia
dds_60 <- dds[, colData(dds)$batch == 60] 
dds_60A <- dds_60[, colData(dds_60)$tissue == "A"]
dds_60Pp <- dds_60[, colData(dds_60)$tissue == "Pp"]

dds_71 <- dds[, colData(dds)$batch == 71] 
dds_71A <- dds_71[, colData(dds_71)$tissue == "A"]
dds_71Pp <- dds_71[, colData(dds_71)$tissue == "Pp"]

### Remove genes counted zero times in subset
dds_60 <- dds_60[ rowSums(counts(dds_60)) > 0, ]
dds_60A <- dds_60A[ rowSums(counts(dds_60A)) > 0, ]
dds_60Pp <- dds_60Pp[ rowSums(counts(dds_60Pp)) > 0, ]

dds_71 <- dds_71[ rowSums(counts(dds_71)) > 0, ]
dds_71A <- dds_71A[ rowSums(counts(dds_71A)) > 0, ]
dds_71Pp <- dds_71Pp[ rowSums(counts(dds_71Pp)) > 0, ]

dds_A <- dds[, colData(dds)$tissue == "A"] 
dds_Pp <- dds[, colData(dds)$tissue == "Pp"] 

### Drop unused factor levels
colData(dds_60A) <- droplevels(colData(dds_60A))
colData(dds_60Pp) <- droplevels(colData(dds_60Pp))

colData(dds_71A) <- droplevels(colData(dds_71A))
colData(dds_71Pp) <- droplevels(colData(dds_71Pp))

colData(dds_A) <- droplevels(colData(dds_A))
colData(dds_Pp) <- droplevels(colData(dds_Pp))
```

### Summarize count data
```{r summarize_counts, fig.height = 4, fig.width = 10}
# Read in count totals of raw, postQC, and mapped reads
count_summ <- read.table("data/count_totals.txt") %>%
  mutate(sample_ID = str_sub(V1, 1, 7),
         raw = V2, post_qc = V3, mapped = V4) %>%
  select(sample_ID, raw, post_qc, mapped) %>%
  gather(stage, count, -sample_ID) 

# Summarize counts for only samples
counttable <- sdat0 %>%
  left_join(count_summ) %>%
  group_by(stage) %>%
  summarize(`Min. (per sample)` = min(count),
            `Max. (per sample)` = max(count),
            `Median (per sample)` = median(count),
            Total = sum(count)) %>%
  arrange(rev(stage))
 
counttable <- counttable[-1,]
counttable %>%
   knitr::kable(caption = "Total read counts")

# Number of samples
nsamples <- ncol(counts(dds))

# Number of reads per sample
rps <- qplot(colSums(counts(dds))) +
  labs(x = "Mapped reads per sample", y = "Number of samples",
       title = "Mapped reads per sample") +
  geom_label(aes(x = 6e5, y = 13, label = paste(nsamples, "samples")))

# Number of genes
ngenes <- nrow(counts(dds))
# Number of reads per gene
rpg <- qplot(log10(rowSums(counts(dds))), bins = 16) +
  labs(x = "log10(Mapped reads per gene)", y = "Number of genes",
       title = "Mapped reads per gene") +
  geom_label(aes(x = 4, y = 1500, label = paste(ngenes, "genes")))

countfig <- plot_grid(rps, rpg)

countfig

# Save table and fig to include in supplemental information
write_csv(counttable, path = "output/readcounttable.csv")
ggsave(countfig, filename = "figures/FigS1_TagSeq_seq_Stats.png", device = "png", width = 150, height = 85, units = "mm")
```

#### Filter and transform count data
```{r prefilter_dds}

# Remove genes with less than 1 mean count across samples
dds <- dds[ rowMeans(counts(dds)) > 1, ]
dds_60 <- dds_60[ rowMeans(counts(dds_60)) > 1, ]
dds_60A <- dds_60A[ rowMeans(counts(dds_60A)) > 1, ]
dds_60Pp <- dds_60Pp[ rowMeans(counts(dds_60Pp)) > 1, ]

dds_71 <- dds_71[ rowMeans(counts(dds_71)) > 1, ]
dds_71A <- dds_71A[ rowMeans(counts(dds_71A)) > 1, ]
dds_71Pp <- dds_71Pp[ rowMeans(counts(dds_71Pp)) > 1, ]

dds_A <- dds_A[ rowMeans(counts(dds_A)) > 1, ]
dds_Pp <- dds_Pp[ rowMeans(counts(dds_Pp)) > 1, ]

# Normalize expression data for visualization purposes using VST tranformation
vsd <- vst(dds, blind = TRUE)
vsd_60 <- vst(dds_60, blind = TRUE)
vsd_60A <- vst(dds_60A, blind = TRUE)
vsd_60Pp <- vst(dds_60Pp, blind = TRUE)

vsd_71 <- vst(dds_71, blind = TRUE)
vsd_71A <- vst(dds_71A, blind = TRUE)
vsd_71Pp <- vst(dds_71Pp, blind = TRUE)

vsd_A <- vst(dds_A, blind = TRUE)
vsd_Pp <- vst(dds_Pp, blind = TRUE)
```

# Visualize transcriptome

### Principal Coordinates Analysis

```{r pcoa, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd)) %>% 
  cbind(cmdscale(sampleDistMatrix)) %>%
  mutate(fillcol = factor(ifelse(tissue == "A", NA, tissue))) %>% 
  mutate(alphaval = ifelse(batch == 60, 1, 0))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(tissue, batch, treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment, shape = tissue)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2),
               lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2, fill = treatment, alpha = as.character(alphaval))) +
  geom_point(size = 2, aes(x = c1, y = c2), fill = ifelse(mds$alphaval, "white", NA)) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`, fill = treatment, alpha = as.character(alphaval)), show.legend = F) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`), fill = ifelse(mds$alphaval, "white", NA), show.legend = F) +
  scale_shape_manual(values = c(21, 24), name = "ganglia", labels = c("abdominal", "pleural & pedal")) +
  scale_alpha_manual(values = c(1, 0), name = "batch", labels = c("lab_reared parents", "wild parents")) +
  guides(shape = guide_legend(order = 2, override.aes = list(fill = "black"))) +
  guides(alpha = guide_legend(override.aes = list(shape = 21, alpha = 1, fill = c("black", "white")))) +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing.y = unit(0, "cm"))

pcoa

ggsave(filename = "figures/Fig1.pcoa_GE.png",pcoa, width = 5, height = 5)
```

The PCoA shows that global gene expression varies significantly by batch along PC1 and by ganglia along PC2.In order to see clear treatment effects I will perform PCoAs for each batch and ganglia

```{r pcoa 60, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_60)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_60)) %>% 
  cbind(cmdscale(sampleDistMatrix)) %>%
  mutate(fillcol = factor(ifelse(tissue == "A", NA, tissue)))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(tissue, treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment, shape = tissue)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_shape_discrete(name = "ganglia", labels = c("abdominal", "pleural & pedal")) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_batch60_GE.png",pcoa, width = 160, height = 110, units = "mm")

```

```{r pcoa 71, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_71)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_71)) %>% 
  cbind(cmdscale(sampleDistMatrix)) %>%
  mutate(fillcol = factor(ifelse(tissue == "A", NA, tissue)))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(tissue, treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment, shape = tissue)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_shape_discrete(name = "ganglia", labels = c("abdominal", "pleural & pedal")) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_batch71_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa Abdominal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_A)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_A)) %>% 
  cbind(cmdscale(sampleDistMatrix)) %>%
  mutate(btch = factor(batch))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(batch, treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment, shape = btch)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_shape_discrete(name = "batch", labels = c("lab_reared parents", "wild parents")) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_Abdominal_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa Pleural & Pedal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_Pp)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_Pp)) %>% 
  cbind(cmdscale(sampleDistMatrix)) %>%
  mutate(btch = factor(batch))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(batch, treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment, shape = btch)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_shape_discrete(name = "batch", labels = c("lab_reared parents", "wild parents")) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_PlePedal_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa batch 60 Abdominal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_60A)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_60A)) %>% 
  cbind(cmdscale(sampleDistMatrix))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_60A_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa batch 60 PlePedal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_60Pp)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_60Pp)) %>% 
  cbind(cmdscale(sampleDistMatrix))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_60Pp_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa batch 71 Abdominal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_71A)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_71A)) %>% 
  cbind(cmdscale(sampleDistMatrix))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_71A_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r pcoa batch 71 PlePedal, eval = TRUE}
## Calculate distances among samples
sampleDists <- dist(t(assay(vsd_71Pp)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)

## Calculate MDS
mds <- as.data.frame(colData(vsd_71Pp)) %>% 
  cbind(cmdscale(sampleDistMatrix))

# Calculate group centroids for plotting
mds <- mds %>%
  group_by(treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)

# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]

# Plot with spiders
pcoa <- ggplot(mds, aes(color = treatment)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2), show.legend = FALSE) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`)) +
  scale_color_discrete(name = "treatment") +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing = unit(0, "cm"), legend.key.size = unit(0.7, 'lines')) +
  guides(fill=guide_legend(ncol=2)) 

pcoa

ggsave(filename = "figures/Fig.pcoa_71Pp_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

### Discriminant Analysis
Discriminant Analysis of Principal Components (DAPC)
```{r DAPC_71, eval = TRUE}

set.seed(1234)
# Discriminant function including all genes

dat <- data.frame(assay(vsd_71)) 

# How many PCs should be kept?
dapc2 <- dapc(t(dat), colData(dds_71)$treatment, n.da=13, n.pca=100)
temp <- optim.a.score(dapc2, n.sim = 50)
my.dapc <- function(n.pca) dapc(t(dat), colData(dds_71)$treatment, n.pca = n.pca, n.da = 13)

library(furrr)
plan(multiprocess)
my.dapc.res <- tibble(n.pca = 2:30) %>%
  mutate(dapc = map(n.pca, my.dapc),
         a.score = furrr::future_map(dapc, a.score, n.sim = 1000, seed = TRUE),
          mean = map_dbl(a.score, ~ .$mean),
          cumvar = map_dbl(dapc, ~ .$var))
 
my.dapc.res %>%
   arrange(-mean) %>%
   head()

# Retaining 8 PC's gives a high a-score (52%) and utilizes 63% of variance. 

dp1 <- dapc(t(dat), colData(dds_71)$treatment,
            n.pca = 7, n.da = 6)   # Retain 7 PCs and 6 discriminant functions
varexpl <- round((dp1$eig/sum(dp1$eig))[1:2] * 100, 1)

dapc <- tibble(sample = rownames(dp1$ind.coord),
               grp = dp1$grp,
               LD1 = dp1$ind.coord[,1],
               LD2 = dp1$ind.coord[,2])
dapc <- dapc %>%
  group_by(grp) %>%
  summarize(c1 = mean(LD1),
            c2 = mean(LD2)) %>%
  full_join(dapc)

# Plot with spiders
dapc.fig <- 
  ggplot(dapc, aes(shape = factor(grepl("Pp", sample)), 
                             fill = factor(grp, levels = c("Control","Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery"), ordered = T))) +
  geom_segment(mapping = aes(x = LD1, y = LD2, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(aes(x = c1, y = c2), size = 2) +
  geom_point(aes(x = LD1, y = LD2), size = 1, show.legend = FALSE) +
  scale_shape_manual(name = "ganglia", labels = c("Abdominal", "Pleural & Pedal"), values = c(21, 24)) +
  scale_fill_discrete(name = "treatment") +
  guides(fill = guide_legend(override.aes = list(shape = 21, size = 2))) +
  guides(shape = guide_legend(override.aes = list(fill = "black", size = 2))) +
  labs(x = paste0("LD1 [", varexpl[1],"%]"), y = paste0("LD2 [", varexpl[2],"%]")) +
  theme_custom() 

dapc.fig

ggsave(filename = "figures/Fig.DAPC_71_GE.png",pcoa, width = 171, height = 110, units = "mm")

```

```{r DAPC_60, eval = TRUE}

set.seed(1234)
# Discriminant function including all genes

dat <- data.frame(assay(vsd_60)) 

# How many PCs should be kept?
dapc2 <- dapc(t(dat), colData(dds_60)$treatment, n.da=13, n.pca=100)
temp <- optim.a.score(dapc2, n.sim = 50)
my.dapc <- function(n.pca) dapc(t(dat), colData(dds_60)$treatment, n.pca = n.pca, n.da = 13)

library(furrr)
plan(multiprocess)
my.dapc.res <- tibble(n.pca = 2:30) %>%
  mutate(dapc = map(n.pca, my.dapc),
         a.score = furrr::future_map(dapc, a.score, n.sim = 1000, seed = TRUE),
          mean = map_dbl(a.score, ~ .$mean),
          cumvar = map_dbl(dapc, ~ .$var))
 
my.dapc.res %>%
   arrange(-mean) %>%
   head()

# Retaining 8 PC's gives a high a-score (52%) and utilizes 63% of variance. 

dp1 <- dapc(t(dat), colData(dds_60)$treatment,
            n.pca = 7, n.da = 6)   # Retain 7 PCs and 6 discriminant functions
varexpl <- round((dp1$eig/sum(dp1$eig))[1:2] * 100, 1)

dapc <- tibble(sample = rownames(dp1$ind.coord),
               grp = dp1$grp,
               LD1 = dp1$ind.coord[,1],
               LD2 = dp1$ind.coord[,2])
dapc <- dapc %>%
  group_by(grp) %>%
  summarize(c1 = mean(LD1),
            c2 = mean(LD2)) %>%
  full_join(dapc)

# Plot with spiders
dapc.fig <- 
  ggplot(dapc, aes(shape = factor(grepl("Pp", sample)), 
                             fill = factor(grp, levels = c("Control","Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery"), ordered = T))) +
  geom_segment(mapping = aes(x = LD1, y = LD2, xend = c1, yend = c2), lwd = 0.25, col = "grey") +
  geom_point(aes(x = c1, y = c2), size = 2) +
  geom_point(aes(x = LD1, y = LD2), size = 1, show.legend = FALSE) +
  scale_shape_manual(name = "ganglia", labels = c("Abdominal", "Pleural & Pedal"), values = c(21, 24)) +
  scale_fill_discrete(name = "treatment") +
  guides(fill = guide_legend(override.aes = list(shape = 21, size = 2))) +
  guides(shape = guide_legend(override.aes = list(fill = "black", size = 2))) +
  labs(x = paste0("LD1 [", varexpl[1],"%]"), y = paste0("LD2 [", varexpl[2],"%]")) +
  theme_custom() 

dapc.fig

ggsave(filename = "figures/Fig.DAPC_60_GE.png",pcoa, width = 171, height = 110, units = "mm")

```
The DAPC shows that despite the significant variability among ganglia, the treatment groups can be discriminated well based on gene expression data and show a consistent response. This is true for both families (60 and 71). 

# Differential expression analysis

Since families and tissues appear to respond (at least on PCoA) differently to hypoxia, we should analyze differential expression in each family and tissue separately to assess their unique responses. We can also analyze differential expression across all tissue and across families together to see what responses are common.

### Run DESeq
Starting with Batch 60 (in-house parents)
First, within each batch and across tissue
```{r run_DESeq_by tissue batch 60, warning=FALSE}
# Run DESeq pipeline
design(dds_60) <- formula(~ tissue.group)
dsr1 <- DESeq(dds_60)

# Define group contrasts
group.contrasts <- tibble(num = c("Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery",
                                  "Hyp_T6h","Reox","Recovery","Reox","LtH_6","LtH_7"),
                          den = c("Control","Control","Control","Control","Control","Control",
                                  "Hyp_T2h","Hyp_T6h","Hyp_T6h","Recovery","Hyp_T6h","Hyp_T6h"))

# Get DESeq results for all group contrasts for each tissue
DE <- crossing(tissue = c("A", "Pp"), group.contrasts) %>%
  mutate(dsr = pmap(list(tissue, num, den), function(x, y, z) {
    results(dsr1, contrast = c("tissue.group", paste0(x, ".", c(y, z))))}))
```

Then, for both tissues together
```{r run_DESeq_together batch 60}
# Run DESeq pipeline 
design(dds_60) <- formula(~ tissue + treatment)
dsr2 <- DESeq(dds_60)

# Get DESeq results for all contrasts and join with results from each colony, from above
DE <- crossing(tissue = "all", group.contrasts) %>%
  mutate(dsr = map2(num, den, ~ results(dsr2, contrast = c("treatment", .x, .y)))) %>%
  bind_rows(DE)
```


### Get significant DEGs
```{r get_DEgenes}
DE <- DE %>%
  mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.1), ]), "gene")),
          up = map(sig, ~ filter(., log2FoldChange > 0)),
        down = map(sig, ~ filter(., log2FoldChange < 0))) 

# Generate logP values for all differential expression contrasts
DE <- DE %>% mutate(logP = map(dsr, ~ data.frame(
  gene = rownames(data.frame(.)),
  logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange))))

# Count number of differentially expressed genes within each tissue, and overall, for each contrast
DEtab <- DE %>%
  filter((num == "Hyp_T2h" & den == "Control")|(num == "Hyp_T6h" & den == "Control")|
         (num == "LtH_6" & den == "Control")|(num == "LtH_7" & den == "Control")|
         (num == "Reox" & den == "Control")|(num == "Recovery" & den == "Control")|
         (num == "Hyp_T6h" & den == "Hyp_T2h")|(num == "Reox" & den == "Hyp_T6h")|
         (num == "Recovery" & den == "Hyp_T6h")|(num == "Reox" & den == "Recovery")|
         (num == "LtH_6" & den == "Hyp_T6h")|(num == "LtH_7" & den == "Hyp_T6h")) %>%
  mutate(nsig = map_dbl( sig, ~ nrow(.)),
          nup = map_dbl(  up, ~ nrow(.)),
          ndn = map_dbl(down, ~ nrow(.)),
          `DEGs [up, down]` = paste0(nsig, " [", nup, ", ", ndn, "]")) %>%
  mutate(ganglia = case_when(tissue == "A" ~ "abdominal", tissue == "Pp" ~ "pleural & pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(ganglia = factor(ganglia, levels = c("abdominal", "pleural & pedal", "all ganglia"))) %>%
  select(ganglia, num, den, `DEGs [up, down]`) %>%
  spread(ganglia, `DEGs [up, down]`)

DEtab$contrast <- paste(DEtab$num, DEtab$den, sep = " vs ")
dek <- DEtab %>% select(c(6,3:5)) %>%
  knitr::kable(format = "markdown", caption = "Number of differentially expressed genes within individual genets and across all genets for each specified contrast. DEGs identified using DESeq with an adjusted p-value < 0.1. In brackets are the numbers of significantly up- and down-regulated genes.", row.names = )

dek
```

Then Batch 71 (wild parents)
```{r run_DESeq_by tissue batch 71, warning=FALSE}
# Run DESeq pipeline
design(dds_71) <- formula(~ tissue.group)
dsr1 <- DESeq(dds_71)

# Define group contrasts
group.contrasts <- tibble(num = c("Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery",
                                  "Hyp_T6h","Reox","Recovery","Reox","LtH_6","LtH_7"),
                          den = c("Control","Control","Control","Control","Control","Control",
                                  "Hyp_T2h","Hyp_T6h","Hyp_T6h","Recovery","Hyp_T6h","Hyp_T6h"))

# Get DESeq results for all group contrasts for each tissue
DE_71 <- crossing(tissue = c("A", "Pp"), group.contrasts) %>%
  mutate(dsr = pmap(list(tissue, num, den), function(x, y, z) {
    results(dsr1, contrast = c("tissue.group", paste0(x, ".", c(y, z))))}))
```

Then, for both tissues together
```{r run_DESeq_together batch 71}
# Run DESeq pipeline 
design(dds_71) <- formula(~ tissue + treatment)
dsr2 <- DESeq(dds_71)

# Get DESeq results for all contrasts and join with results from each tissue, from above
DE_71 <- crossing(tissue = "all", group.contrasts) %>%
  mutate(dsr = map2(num, den, ~ results(dsr2, contrast = c("treatment", .x, .y)))) %>%
  bind_rows(DE_71)
```


### Get significant DEGs
```{r get_DEgenes batch_71}
DE_71 <- DE_71 %>%
  mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.1), ]), "gene")),
          up = map(sig, ~ filter(., log2FoldChange > 0)),
        down = map(sig, ~ filter(., log2FoldChange < 0))) 

# Generate logP values for all differential expression contrasts
DE_71 <- DE_71 %>% mutate(logP = map(dsr, ~ data.frame(
  gene = rownames(data.frame(.)),
  logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange))))

# Count number of differentially expressed genes within each tissue, and overall, for each contrast
DEtab_71 <- DE_71 %>%
  filter((num == "Hyp_T2h" & den == "Control")|(num == "Hyp_T6h" & den == "Control")|
         (num == "LtH_6" & den == "Control")|(num == "LtH_7" & den == "Control")|
         (num == "Reox" & den == "Control")|(num == "Recovery" & den == "Control")|
         (num == "Hyp_T6h" & den == "Hyp_T2h")|(num == "Reox" & den == "Hyp_T6h")|
         (num == "Recovery" & den == "Hyp_T6h")|(num == "Reox" & den == "Recovery")|
         (num == "LtH_6" & den == "Hyp_T6h")|(num == "LtH_7" & den == "Hyp_T6h")) %>%
  mutate(nsig = map_dbl( sig, ~ nrow(.)),
          nup = map_dbl(  up, ~ nrow(.)),
          ndn = map_dbl(down, ~ nrow(.)),
          `DEGs [up, down]` = paste0(nsig, " [", nup, ", ", ndn, "]")) %>%
  mutate(ganglia = case_when(tissue == "A" ~ "abdominal", tissue == "Pp" ~ "pleural & pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(ganglia = factor(ganglia, levels = c("abdominal", "pleural & pedal", "all ganglia"))) %>%
  select(ganglia, num, den, `DEGs [up, down]`) %>%
  spread(ganglia, `DEGs [up, down]`)

DEtab_71$contrast <- paste(DEtab_71$num, DEtab$den, sep = " vs ")
dek <- DEtab_71 %>% select(c(6,3:5)) %>%
  knitr::kable(format = "markdown", caption = "Number of differentially expressed genes within individual genets and across all genets for each specified contrast. DEGs identified using DESeq with an adjusted p-value < 0.1. In brackets are the numbers of significantly up- and down-regulated genes.", row.names = )

dek
```

Now within each tissue and across batches
```{r run_DESeq_for Abdominal, warning=FALSE}
# Run DESeq pipeline
design(dds_A) <- formula(~ batch.group)
dsr1 <- DESeq(dds_A)

# Define group contrasts
group.contrasts <- tibble(num = c("Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery",
                                  "Hyp_T6h","Reox","Recovery","Reox","LtH_6","LtH_7"),
                          den = c("Control","Control","Control","Control","Control","Control",
                                  "Hyp_T2h","Hyp_T6h","Hyp_T6h","Recovery","Hyp_T6h","Hyp_T6h"))

# Get DESeq results for all group contrasts for each batch
DE_A <- crossing(batch = c("60", "71"), group.contrasts) %>%
  mutate(dsr = pmap(list(batch, num, den), function(x, y, z) {
    results(dsr1, contrast = c("batch.group", paste0(x, ".", c(y, z))))}))
```

Then, for both batches together
```{r run_DESeq_together abdominal}
# Run DESeq pipeline 
design(dds_A) <- formula(~ batch + treatment)
dsr2 <- DESeq(dds_A)

# Get DESeq results for all contrasts and join with results from each batch, from above
DE_A <- crossing(batch = "all", group.contrasts) %>%
  mutate(dsr = map2(num, den, ~ results(dsr2, contrast = c("treatment", .x, .y)))) %>%
  bind_rows(DE_A)
```


### Get significant DEGs
```{r get_DEgenes Abdominal}
DE_A <- DE_A %>%
  mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.1), ]), "gene")),
          up = map(sig, ~ filter(., log2FoldChange > 0)),
        down = map(sig, ~ filter(., log2FoldChange < 0))) 

# Generate logP values for all differential expression contrasts
DE_A <- DE_A %>% mutate(logP = map(dsr, ~ data.frame(
  gene = rownames(data.frame(.)),
  logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange))))

# Count number of differentially expressed genes within each batch, and overall, for each contrast
DE_Atab <- DE_A %>%
  filter((num == "Hyp_T2h" & den == "Control")|(num == "Hyp_T6h" & den == "Control")|
         (num == "LtH_6" & den == "Control")|(num == "LtH_7" & den == "Control")|
         (num == "Reox" & den == "Control")|(num == "Recovery" & den == "Control")|
         (num == "Hyp_T6h" & den == "Hyp_T2h")|(num == "Reox" & den == "Hyp_T6h")|
         (num == "Recovery" & den == "Hyp_T6h")|(num == "Reox" & den == "Recovery")|
         (num == "LtH_6" & den == "Hyp_T6h")|(num == "LtH_7" & den == "Hyp_T6h")) %>%
  mutate(nsig = map_dbl( sig, ~ nrow(.)),
          nup = map_dbl(  up, ~ nrow(.)),
          ndn = map_dbl(down, ~ nrow(.)),
          `DEGs [up, down]` = paste0(nsig, " [", nup, ", ", ndn, "]")) %>%
  mutate(family = case_when(batch == "60" ~ "lab parents", batch == "71" ~ "wild parents", 
                           batch == "all" ~ "all batches")) %>%
  mutate(family = factor(family, levels = c("lab parents", "wild parents", "all batches"))) %>%
  select(family, num, den, `DEGs [up, down]`) %>%
  spread(family, `DEGs [up, down]`)

DE_Atab$contrast <- paste(DE_Atab$num, DE_Atab$den, sep = " vs ")
dek <- DE_Atab %>% select(c(6,3:5)) %>%
  knitr::kable(format = "markdown", caption = "Number of differentially expressed genes within individual genets and across all genets for each specified contrast. DEGs identified using DESeq with an adjusted p-value < 0.1. In brackets are the numbers of significantly up- and down-regulated genes.", row.names = )

dek
```

```{r run_DESeq_for Pleural/Pedal, warning=FALSE}
# Run DESeq pipeline
design(dds_Pp) <- formula(~ batch.group)
dsr1 <- DESeq(dds_Pp)

# Define group contrasts
group.contrasts <- tibble(num = c("Hyp_T2h","Hyp_T6h","LtH_6","LtH_7","Reox","Recovery",
                                  "Hyp_T6h","Reox","Recovery","Reox","LtH_6","LtH_7"),
                          den = c("Control","Control","Control","Control","Control","Control",
                                  "Hyp_T2h","Hyp_T6h","Hyp_T6h","Recovery","Hyp_T6h","Hyp_T6h"))

# Get DESeq results for all group contrasts for each batch
DE_Pp <- crossing(batch = c("60", "71"), group.contrasts) %>%
  mutate(dsr = pmap(list(batch, num, den), function(x, y, z) {
    results(dsr1, contrast = c("batch.group", paste0(x, ".", c(y, z))))}))
```

Then, for both batches together
```{r run_DESeq_together Pleural & Pedal}
# Run DESeq pipeline 
design(dds_Pp) <- formula(~ batch + treatment)
dsr2 <- DESeq(dds_Pp)

# Get DESeq results for all contrasts and join with results from each batch, from above
DE_Pp <- crossing(batch = "all", group.contrasts) %>%
  mutate(dsr = map2(num, den, ~ results(dsr2, contrast = c("treatment", .x, .y)))) %>%
  bind_rows(DE_Pp)
```


### Get significant DEGs
```{r get_DEgenes Pleural & Pedal}
DE_Pp <- DE_Pp %>%
  mutate(sig = map(dsr, ~ rownames_to_column(data.frame(.[which(.$padj < 0.05), ]), "gene")),
          up = map(sig, ~ filter(., log2FoldChange > 0)),
        down = map(sig, ~ filter(., log2FoldChange < 0))) 

# Generate logP values for all differential expression contrasts
DE_Pp <- DE_Pp %>% mutate(logP = map(dsr, ~ data.frame(
  gene = rownames(data.frame(.)),
  logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange))))

# Count number of differentially expressed genes within each batch, and overall, for each contrast
DE_Pptab <- DE_Pp %>%
  filter((num == "Hyp_T2h" & den == "Control")|(num == "Hyp_T6h" & den == "Control")|
         (num == "LtH_6" & den == "Control")|(num == "LtH_7" & den == "Control")|
         (num == "Reox" & den == "Control")|(num == "Recovery" & den == "Control")|
         (num == "Hyp_T6h" & den == "Hyp_T2h")|(num == "Reox" & den == "Hyp_T6h")|
         (num == "Recovery" & den == "Hyp_T6h")|(num == "Reox" & den == "Recovery")|
         (num == "LtH_6" & den == "Hyp_T6h")|(num == "LtH_7" & den == "Hyp_T6h")) %>%
  mutate(nsig = map_dbl( sig, ~ nrow(.)),
          nup = map_dbl(  up, ~ nrow(.)),
          ndn = map_dbl(down, ~ nrow(.)),
          `DEGs [up, down]` = paste0(nsig, " [", nup, ", ", ndn, "]")) %>%
  mutate(family = case_when(batch == "60" ~ "lab parents", batch == "71" ~ "wild parents", 
                           batch == "all" ~ "all batches")) %>%
  mutate(family = factor(family, levels = c("lab parents", "wild parents", "all batches"))) %>%
  select(family, num, den, `DEGs [up, down]`) %>%
  spread(family, `DEGs [up, down]`)

DE_Pptab$contrast <- paste(DE_Pptab$num, DE_Pptab$den, sep = " vs ")
dek <- DE_Pptab %>% select(c(6,3:5)) %>%
  knitr::kable(format = "markdown", caption = "Number of differentially expressed genes within individual genets and across all genets for each specified contrast. DEGs identified using DESeq with an adjusted p-value < 0.1. In brackets are the numbers of significantly up- and down-regulated genes.", row.names = )

dek
```

```{r graph DEGs}

DEGs_60A <- DE %>%
  unnest(sig) %>%
  filter(tissue == "A") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene)

DEGs_60A_wide <- list(Hyp_T2h=pull(DEGs_60A[which(DEGs_60A$contrast=="Hyp_T2h.Control"),2]),
                 Hyp_T6h=pull(DEGs_60A[which(DEGs_60A$contrast=="Hyp_T6h.Control"),2]),
                 Hyp_T6d=pull(DEGs_60A[which(DEGs_60A$contrast=="LtH_6.Control"),2]),
                 Hyp_T7d=pull(DEGs_60A[which(DEGs_60A$contrast=="LtH_7.Control"),2]))
str(DEGs_60A_wide)
gv1 <- ggvenn(DEGs_60A_wide, fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv1
ggsave("figures/A_60_DEGs_vennDiagram.png", gv1, height = 3, width = 5)

DEGs_60Pp <- DE %>%
  unnest(sig) %>%
  filter(tissue == "Pp") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene)

DEGs_60Pp_wide <- list(Hyp_T2h=pull(DEGs_60Pp[which(DEGs_60Pp$contrast=="Hyp_T2h.Control"),2]),
                 Hyp_T6h=pull(DEGs_60Pp[which(DEGs_60Pp$contrast=="Hyp_T6h.Control"),2]),
                 Hyp_T6d=pull(DEGs_60Pp[which(DEGs_60Pp$contrast=="LtH_6.Control"),2]),
                 Hyp_T7d=pull(DEGs_60Pp[which(DEGs_60Pp$contrast=="LtH_7.Control"),2]))
str(DEGs_60Pp_wide)
gv2 <- ggvenn(DEGs_60Pp_wide, fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv2
ggsave("figures/Pp_60_DEGs_vennDiagram.png", gv2, height = 3, width = 5)

DEGs_71A <- DE_71 %>%
  unnest(sig) %>%
  filter(tissue == "A") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene)
DEGs_71A_wide <- list(Hyp_T2h=pull(DEGs_71A[which(DEGs_71A$contrast=="Hyp_T2h.Control"),2]),
                 Hyp_T6h=pull(DEGs_71A[which(DEGs_71A$contrast=="Hyp_T6h.Control"),2]),
                 Hyp_T6d=pull(DEGs_71A[which(DEGs_71A$contrast=="LtH_6.Control"),2]),
                 Hyp_T7d=pull(DEGs_71A[which(DEGs_71A$contrast=="LtH_7.Control"),2]))
str(DEGs_71A_wide)
gv3 <- ggvenn(DEGs_71A_wide, fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv3
ggsave("figures/A_71_DEGs_vennDiagram.png", gv3, height = 3, width = 5)

DEGs_71Pp <- DE_71 %>%
  unnest(sig) %>%
  filter(tissue == "Pp") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene)
DEGs_71Pp_wide <- list(Hyp_T2h=pull(DEGs_71Pp[which(DEGs_71Pp$contrast=="Hyp_T2h.Control"),2]),
                 Hyp_T6h=pull(DEGs_71Pp[which(DEGs_71Pp$contrast=="Hyp_T6h.Control"),2]),
                 Hyp_T6d=pull(DEGs_71Pp[which(DEGs_71Pp$contrast=="LtH_6.Control"),2]),
                 Hyp_T7d=pull(DEGs_71Pp[which(DEGs_71Pp$contrast=="LtH_7.Control"),2]))
str(DEGs_71Pp_wide)
gv4 <- ggvenn(DEGs_71Pp_wide, fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv4
ggsave("figures/Pp_71_DEGs_vennDiagram.png", gv4, height = 3, width = 5)

DEGs_A <- DE_A %>%
  unnest(sig) %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  filter(contrast %in% c("Hyp_T2h.Control", "Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  select(.,contrast, gene, batch)
DEGs_A_wide <- list(wild=pull(DEGs_A[which(DEGs_A$batch==71),2]),
                 lab_reared=pull(DEGs_A[which(DEGs_A$batch==60),2]))
str(DEGs_A_wide)
gv5 <- ggvenn(DEGs_A_wide, fill_color = c("#0073C2FF", "#EFC000FF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv5
ggsave("figures/A_DEGs_vennDiagram.png", gv5, height = 3, width = 5)

DEGs_Pp <- DE_Pp %>%
  unnest(sig) %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  filter(contrast %in% c("Hyp_T2h.Control", "Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  select(.,contrast, gene, batch)
DEGs_Pp_wide <- list(wild=pull(DEGs_Pp[which(DEGs_Pp$batch==71),2]),
                 lab_reared=pull(DEGs_Pp[which(DEGs_Pp$batch==60),2]))
str(DEGs_Pp_wide)
gv6 <- ggvenn(DEGs_Pp_wide, fill_color = c("#0073C2FF", "#EFC000FF"),
  stroke_size = 0.5, set_name_size = 4, show_percentage = F, stroke_color = F 
  )
gv6
ggsave("figures/Pp_DEGs_vennDiagram.png", gv6, height = 3, width = 5)
```

```{r}
library(VennDiagram)
library(RColorBrewer)
Col <- brewer.pal(8, "Pastel1")
myCol <- Col[1:5]
venn.diagram(
        x = DEGs_60A_wide,
        category.names = c("Hyp_2h", "Hyp_6h", "Hyp_6d", "Hyp_7d","Recovery"),
        filename = 'figures/DMGs_60A_venn_diagramm.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans"
)
##Create Venn Diagram of DCvCC
myCol1 <- c("#CCEBC5","#DECBE4")
venn.diagram(
        x = list(gene_list_DEG_DCvCC,gene_list_DMG_DCvCC),
        category.names = c("DEGs", "DMGs"),
        filename = '../output/figures/DMG_DEG_DcCc_venn_diagramm.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol1,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans"
)
##Create Venn Diagram of DHvDC
myCol2 <- c("#FED9A6","#FFFFCC")
venn.diagram(
        x = list(gene_list_DEG_DHvDC,gene_list_DMG_DHvDC),
        category.names = c("DEGs", "DMGs"),
        filename = '../output/figures/DMG_DEG_DhDc_venn_diagramm.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol2,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans"
)
```


### DEG Similarities
Similarity of differentially expressed genes across tissue and contrasts

##### DEGs in common across tissues and batches
```{r DEgenes_comparison}
# Find genes that are DE in same direction in both tissues when analyzed individually
DE %>%
  unnest(sig) %>%
  filter(tissue != "all") %>%
  group_by(num, den) %>%
  dplyr::count(gene, log2FoldChange > 0) %>%   # In how many genets is same gene DE in same direction?
  summarise(`shared` = sum(n==2)) %>%
  knitr::kable(caption = "Number of differentially expressed genes in common between abdominal and Pleural/Pedal ganglia (Batch 60)")  %>%
  kable_styling(full_width = TRUE)

DE_71 %>%
  unnest(sig) %>%
  filter(tissue != "all") %>%
  group_by(num, den) %>%
  dplyr::count(gene, log2FoldChange > 0) %>%   # In how many genets is same gene DE in same direction?
  summarise(`shared` = sum(n==2)) %>%
  knitr::kable(caption = "Number of differentially expressed genes in common between abdominal and Pleural/Pedal ganglia (Batch 71)")  %>%
  kable_styling(full_width = TRUE)

# Find genes that are DE in same direction in both batches when analyzed individually
DE_A %>%
  unnest(sig) %>%
  filter(batch != "all") %>%
  group_by(num, den) %>%
  dplyr::count(gene, log2FoldChange > 0) %>%   # In how many genets is same gene DE in same direction?
  summarise(`shared` = sum(n==2)) %>%
  knitr::kable(caption = "Number of differentially expressed genes in common between batches 60 and 71 (Abdominal ganglium)")  %>%
  kable_styling(full_width = TRUE)

DE_Pp %>%
  unnest(sig) %>%
  filter(batch != "all") %>%
  group_by(num, den) %>%
  dplyr::count(gene, log2FoldChange > 0) %>%   # In how many genets is same gene DE in same direction?
  summarise(`shared` = sum(n==2)) %>%
  knitr::kable(caption = "Number of differentially expressed genes in common between batches 60 and 71 (Pleural/Pedal)")  %>%
  kable_styling(full_width = TRUE)
```

```{r DEgenes_save}

DEGs_60 <- DE %>%
  unnest(sig) %>%
  filter(tissue == "all") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_60, path = "output/DEGs_by_contrast_batch 60_all_ganglia.tsv")

DEGs_60A <- DE %>%
  unnest(sig) %>%
  filter(tissue == "A") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_60A, path = "output/DEGs_by_contrast_batch 60_abdominal.tsv")

DEGs_60Pp <- DE %>%
  unnest(sig) %>%
  filter(tissue == "Pp") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_60Pp, path = "output/DEGs_by_contrast_batch 60_PlePed.tsv")

DEGs_71 <- DE_71 %>%
  unnest(sig) %>%
  filter(tissue == "all") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_71, path = "output/DEGs_by_contrast_batch 71_all_ganglia.tsv")

DEGs_71A <- DE_71 %>%
  unnest(sig) %>%
  filter(tissue == "A") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_71A, path = "output/DEGs_by_contrast_batch 71_abdominal.tsv")

DEGs_71Pp <- DE_71 %>%
  unnest(sig) %>%
  filter(tissue == "Pp") %>%
  mutate(contrast = paste(num, den, sep = ".")) %>%
  select(.,contrast, gene, log2FoldChange)
write_tsv(DEGs_71Pp, path = "output/DEGs_by_contrast_batch 71_PlePed.tsv")

  genes_of_interest <- c("LOC101854685","LOC101860167","LOC100533275","LOC101852344","LOC101855344","SOD")
DEGs_60[which(DEGs_60$gene %in% genes_of_interest),]
DEGs_60A[which(DEGs_60A$gene %in% genes_of_interest),]
DEGs_60Pp[which(DEGs_60Pp$gene %in% genes_of_interest),]
DEGs_71[which(DEGs_71$gene %in% genes_of_interest),]
DEGs_71A[which(DEGs_71A$gene %in% genes_of_interest),]
DEGs_71Pp[which(DEGs_71Pp$gene %in% genes_of_interest),]
```

#### DEGs in common beteween peak of exposure vs Ctrl contrasts (T6h vs Ctrol, LtH_6 vs Ctrl & LtH_7 vs Ctrl) 
in batch 60 abdominal
```{r DEGs peak 60A, fig.height = 8, fig.width = 8}
genesincommon_60A <- DE %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(tissue == "A", contrast %in% c("Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control","Recovery.Control","Reox.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))

genes_of_interest <- c("LOC101854685","LOC101860167","LOC100533275","LOC101852344","LOC101855344")
genesincommon_60A[which(genesincommon_60A$gene %in% genes_of_interest),]  


genesincommon_60A <- genesincommon_60A[c(1:100),]  
# Get normalized counts for the genes in common
gcounts <- data.frame(assay(vsd_60[which(rownames(assay(vsd_60)) %in% genesincommon_60A_int$gene), ]))
gcounts <- rownames_to_column(gcounts, var = "gene") %>%
  pivot_longer(-gene, names_to = "sample_ID", values_to = "count") %>%
  left_join(select(sdat0, sample_ID, treatment)) %>%
  mutate(gene = factor(gene, levels = levels(genesincommon_60A_int$gene))) %>%
  filter(treatment != "Control") %>%
  filter(treatment != "Hyp_T2h") %>%
  filter(treatment != "Reox") %>%
  filter(treatment != "Recovery")


# Get group log2foldchanges for plotting
gl2fc <- genesincommon_60A_int %>%
  pivot_longer(-gene, names_to = "treatment", values_to = "l2fc") %>%
  mutate(treatment = gsub(".Control", "", treatment))
  
# Get group means for plotting
gcountsmean <- gcounts %>%
  group_by(gene, treatment) %>%
  summarise(count = mean(count))

# Plot
plotgcounts <- gcounts %>%
  ggplot(aes(x = treatment, y = pmax(0,log(count)))) +
  geom_jitter(width = 0.2, aes(color = treatment)) +
  geom_line(data = gcountsmean, aes(group = gene)) +
  geom_text(data = gl2fc, aes(y = 2.4, label = round(l2fc, 2))) +
  facet_wrap(~ gene) +
  theme_custom() +
  theme(legend.position = "none") +
  labs(x = "", y = "log(Variance stabilized counts)") +
  scale_y_continuous(limits = c(0, 2.6))
plotgcounts

ggsave(file = "figures/FigS_shared_genes_60A.png", width = 360, height = 360, units = "mm")

```

in batch 60 Pleural/Pedal
```{r DEGs peak 60Pp, fig.height = 5, fig.width = 5}
genesincommon_60Pp <- DE %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(tissue == "Pp", contrast %in% c("Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))  

genesincommon_60Pp_int <- genesincommon_60Pp[which(genesincommon_60Pp$gene %in% genes_of_interest),] 

# Get normalized counts for the genes in common
gcounts <- data.frame(assay(vsd_60[which(rownames(assay(vsd_60)) %in% genesincommon_60Pp$gene), ]))
gcounts <- rownames_to_column(gcounts, var = "gene") %>%
  pivot_longer(-gene, names_to = "sample_ID", values_to = "count") %>%
  left_join(select(sdat0, sample_ID, treatment)) %>%
  mutate(gene = factor(gene, levels = levels(genesincommon_60Pp$gene))) %>%
  filter(treatment != "Control") %>%
  filter(treatment != "Hyp_T2h") %>%
  filter(treatment != "Reox") %>%
  filter(treatment != "Recovery")


# Get group log2foldchanges for plotting
gl2fc <- genesincommon_60Pp %>%
  pivot_longer(-gene, names_to = "treatment", values_to = "l2fc") %>%
  mutate(treatment = gsub(".Control", "", treatment))
  
# Get group means for plotting
gcountsmean <- gcounts %>%
  group_by(gene, treatment) %>%
  summarise(count = mean(count))

# Plot
plotgcounts <- gcounts %>%
  ggplot(aes(x = treatment, y = pmax(0,log(count)))) +
  geom_jitter(width = 0.2, aes(color = treatment)) +
  geom_line(data = gcountsmean, aes(group = gene)) +
  geom_text(data = gl2fc, aes(y = 2.4, label = round(l2fc, 2))) +
  facet_wrap(~ gene) +
  theme_custom() +
  theme(legend.position = "none") +
  labs(x = "", y = "log(Variance stabilized counts)") +
  scale_y_continuous(limits = c(0, 2.6))
plotgcounts

ggsave(file = "figures/FigS_shared_genes_60Pp.png", width = 360, height = 360, units = "mm")

```

in batch 71 abdominal
```{r DEGs peak 71A, fig.height = 8, fig.width = 8}
genesincommon_71A <- DE_71 %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(tissue == "A", contrast %in% c("Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))

genesincommon_71A_int <- genesincommon_71A[which(genesincommon_71A$gene %in% genes_of_interest),] 

genesincommon_71A <- genesincommon_71A[c(1:100),]  
# Get normalized counts for the genes in common
gcounts <- data.frame(assay(vsd_71[which(rownames(assay(vsd_71)) %in% genesincommon_71A$gene), ]))
gcounts <- rownames_to_column(gcounts, var = "gene") %>%
  pivot_longer(-gene, names_to = "sample_ID", values_to = "count") %>%
  left_join(select(sdat0, sample_ID, treatment)) %>%
  mutate(gene = factor(gene, levels = levels(genesincommon_71A$gene))) %>%
  filter(treatment != "Control") %>%
  filter(treatment != "Hyp_T2h") %>%
  filter(treatment != "Reox") %>%
  filter(treatment != "Recovery")


# Get group log2foldchanges for plotting
gl2fc <- genesincommon_71A %>%
  pivot_longer(-gene, names_to = "treatment", values_to = "l2fc") %>%
  mutate(treatment = gsub(".Control", "", treatment))
  
# Get group means for plotting
gcountsmean <- gcounts %>%
  group_by(gene, treatment) %>%
  summarise(count = mean(count))

# Plot
plotgcounts <- gcounts %>%
  ggplot(aes(x = treatment, y = pmax(0,log(count)))) +
  geom_jitter(width = 0.2, aes(color = treatment)) +
  geom_line(data = gcountsmean, aes(group = gene)) +
  geom_text(data = gl2fc, aes(y = 2.4, label = round(l2fc, 2))) +
  facet_wrap(~ gene) +
  theme_custom() +
  theme(legend.position = "none") +
  labs(x = "", y = "log(Variance stabilized counts)") +
  scale_y_continuous(limits = c(0, 2.6))
plotgcounts

ggsave(file = "figures/FigS_shared_genes_71A.png", width = 360, height = 360, units = "mm")

```

in batch 71 Pleural/Pedal
```{r DEGs peak 71Pp, fig.height = 5, fig.width = 5}
genesincommon_71Pp <- DE_71 %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(tissue == "Pp", contrast %in% c("Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))  

genesincommon_71Pp_int <- genesincommon_71Pp[which(genesincommon_71Pp$gene %in% genes_of_interest),] 

# Get normalized counts for the genes in common
gcounts <- data.frame(assay(vsd_71[which(rownames(assay(vsd_71)) %in% genesincommon_71Pp$gene), ]))
gcounts <- rownames_to_column(gcounts, var = "gene") %>%
  pivot_longer(-gene, names_to = "sample_ID", values_to = "count") %>%
  left_join(select(sdat0, sample_ID, treatment)) %>%
  mutate(gene = factor(gene, levels = levels(genesincommon_71Pp$gene))) %>%
  filter(treatment != "Control") %>%
  filter(treatment != "Hyp_T2h") %>%
  filter(treatment != "Reox") %>%
  filter(treatment != "Recovery")


# Get group log2foldchanges for plotting
gl2fc <- genesincommon_71Pp %>%
  pivot_longer(-gene, names_to = "treatment", values_to = "l2fc") %>%
  mutate(treatment = gsub(".Control", "", treatment))
  
# Get group means for plotting
gcountsmean <- gcounts %>%
  group_by(gene, treatment) %>%
  summarise(count = mean(count))

# Plot
plotgcounts <- gcounts %>%
  ggplot(aes(x = treatment, y = pmax(0,log(count)))) +
  geom_jitter(width = 0.2, aes(color = treatment)) +
  geom_line(data = gcountsmean, aes(group = gene)) +
  geom_text(data = gl2fc, aes(y = 2.4, label = round(l2fc, 2))) +
  facet_wrap(~ gene) +
  theme_custom() +
  theme(legend.position = "none") +
  labs(x = "", y = "log(Variance stabilized counts)") +
  scale_y_continuous(limits = c(0, 2.6))
plotgcounts

ggsave(file = "figures/FigS_shared_genes_71Pp.png", width = 360, height = 360, units = "mm")

```

# GO enrichment analysis

```{r GO_MWU contrasts abdominal Cvs2h}
# Start with abdominal 
# Control vs 2h contrast (Rapid response)
setwd("analyses/GO_MWU/")

# Write all-batches control vs. 2h hypoxia logP values to file for GO_MWU analysis
DE_A %>%
  filter(batch == "all", num == "Hyp_T2h", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "A_2h.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("A_2h.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("A_2h.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("A_2h.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
A_Hyp2h.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_A_2h.vs.Ctrl.logP.txt",
    MF = "MWU_MF_A_2h.vs.Ctrl.logP.txt",
    CC = "MWU_CC_A_2h.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_A_2h.vs.Ctrl.logP.txt",
    MF = "MF_A_2h.vs.Ctrl.logP.txt",
    CC = "CC_A_2h.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
A_Hyp2h.Ctrl.GO <- A_Hyp2h.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
A_Hyp2h.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts abdominal Cvs6h}

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 6h hypoxia logP values to file for GO_MWU analysis
#DE_A %>%
  #filter(batch == "all", num == "Hyp_T6h", den == "Control") %>%
  #unnest(logP) %>%
  #select(gene, logP) %>%
  #write.table(., file = "A_6h.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for biological Process (BP) GO terms
commandArgs <- function(...) c("A_6h.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("A_6h.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("A_6h.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
A_Hyp6h.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_A_6h.vs.Ctrl.logP.txt",
    MF = "MWU_MF_A_6h.vs.Ctrl.logP.txt",
    CC = "MWU_CC_A_6h.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_A_6h.vs.Ctrl.logP.txt",
    MF = "MF_A_6h.vs.Ctrl.logP.txt",
    CC = "CC_A_6h.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
A_Hyp6h.Ctrl.GO <- A_Hyp6h.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
A_Hyp6h.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts abdominal Cvs6d}
# Start with abdominal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 6days repeated hypoxia logP values to file for GO_MWU analysis
DE_A %>%
  filter(batch == "all", num == "LtH_6", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "A_6d.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("A_6d.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("A_6d.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("A_6d.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
A_Hyp6d.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_A_6d.vs.Ctrl.logP.txt",
    MF = "MWU_MF_A_6d.vs.Ctrl.logP.txt",
    CC = "MWU_CC_A_6d.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_A_6d.vs.Ctrl.logP.txt",
    MF = "MF_A_6d.vs.Ctrl.logP.txt",
    CC = "CC_A_6d.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
A_Hyp6d.Ctrl.GO <- A_Hyp6d.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
A_Hyp6d.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts abdominal Cvs7d}
# Start with abdominal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 7d hypoxia logP values to file for GO_MWU analysis
DE_A %>%
  filter(batch == "all", num == "LtH_7", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "A_7d.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("A_7d.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("A_7d.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("A_7d.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
A_Hyp7d.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_A_7d.vs.Ctrl.logP.txt",
    MF = "MWU_MF_A_7d.vs.Ctrl.logP.txt",
    CC = "MWU_CC_A_7d.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_A_7d.vs.Ctrl.logP.txt",
    MF = "MF_A_7d.vs.Ctrl.logP.txt",
    CC = "CC_A_7d.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
A_Hyp7d.Ctrl.GO <- A_Hyp7d.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
A_Hyp7d.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts abdominal CvsRecovery}
# Start with abdominal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. Recovery hypoxia logP values to file for GO_MWU analysis
DE_A %>%
  filter(batch == "all", num == "Recovery", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "A_Recovery.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("A_Recovery.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("A_Recovery.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("A_Recovery.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
setwd("analyses/GO_MWU/")
# Get list of all significant GO terms for contrast
A_HypRecovery.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_A_Recovery.vs.Ctrl.logP.txt",
    MF = "MWU_MF_A_Recovery.vs.Ctrl.logP.txt",
    CC = "MWU_CC_A_Recovery.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_A_Recovery.vs.Ctrl.logP.txt",
    MF = "MF_A_Recovery.vs.Ctrl.logP.txt",
    CC = "CC_A_Recovery.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
A_HypRecovery.Ctrl.GO <- A_HypRecovery.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
A_HypRecovery.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts pleural/pedal Cvs2h}
# Start with pleural/pedal 
# Control vs 2h contrast (Rapid response)
setwd("analyses/GO_MWU/")

# Write all-batches control vs. 2h hypoxia logP values to file for GO_MWU analysis
DE_Pp %>%
  filter(batch == "all", num == "Hyp_T2h", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "P_2h.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("P_2h.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("P_2h.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("P_2h.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
P_Hyp2h.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_P_2h.vs.Ctrl.logP.txt",
    MF = "MWU_MF_P_2h.vs.Ctrl.logP.txt",
    CC = "MWU_CC_P_2h.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_P_2h.vs.Ctrl.logP.txt",
    MF = "MF_P_2h.vs.Ctrl.logP.txt",
    CC = "CC_P_2h.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
P_Hyp2h.Ctrl.GO <- P_Hyp2h.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
P_Hyp2h.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts pleural/pedal Cvs6h}

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 6h hypoxia logP values to file for GO_MWU analysis
DE_Pp %>%
  filter(batch == "all", num == "Hyp_T6h", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "P_6h.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("P_6h.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("P_6h.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("P_6h.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
setwd("analyses/GO_MWU/")
# Get list of all significant GO terms for contrast
P_Hyp6h.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_P_6h.vs.Ctrl.logP.txt",
    MF = "MWU_MF_P_6h.vs.Ctrl.logP.txt",
    CC = "MWU_CC_P_6h.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_P_6h.vs.Ctrl.logP.txt",
    MF = "MF_P_6h.vs.Ctrl.logP.txt",
    CC = "CC_P_6h.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
P_Hyp6h.Ctrl.GO <- P_Hyp6h.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
P_Hyp6h.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts pleural/pedal 6dvsCtrl}
# Start with pleural/pedal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 6days repeated hypoxia logP values to file for GO_MWU analysis
DE_Pp %>%
  filter(batch == "all", num == "LtH_6", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "P_6d.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("P_6d.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("P_6d.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("P_6d.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
P_Hyp6d.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_P_6d.vs.Ctrl.logP.txt",
    MF = "MWU_MF_P_6d.vs.Ctrl.logP.txt",
    CC = "MWU_CC_P_6d.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_P_6d.vs.Ctrl.logP.txt",
    MF = "MF_P_6d.vs.Ctrl.logP.txt",
    CC = "CC_P_6d.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
P_Hyp6d.Ctrl.GO <- P_Hyp6d.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
P_Hyp6d.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts pleural/pedal 7dvsCtrl}
# Start with pleural/pedal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. 7d hypoxia logP values to file for GO_MWU analysis
DE_Pp %>%
  filter(batch == "all", num == "LtH_7", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "P_7d.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("P_7d.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("P_7d.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("P_7d.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")
# Get list of all significant GO terms for contrast
P_Hyp7d.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_P_7d.vs.Ctrl.logP.txt",
    MF = "MWU_MF_P_7d.vs.Ctrl.logP.txt",
    CC = "MWU_CC_P_7d.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_P_7d.vs.Ctrl.logP.txt",
    MF = "MF_P_7d.vs.Ctrl.logP.txt",
    CC = "CC_P_7d.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
P_Hyp7d.Ctrl.GO <- P_Hyp7d.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
P_Hyp7d.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```

```{r GO_MWU contrasts pleural/pedal RecoveryVsCtrol}
# Start with pleural/pedal 

setwd("analyses/GO_MWU/")

# Write all-batches control vs. Recovery hypoxia logP values to file for GO_MWU analysis
DE_Pp %>%
  filter(batch == "all", num == "Recovery", den == "Control") %>%
  unnest(logP) %>%
  select(gene, logP) %>%
  write.table(., file = "P_Recovery.vs.Ctrl.logP.txt", row.names = FALSE, quote = FALSE, sep = ",")

# Run GO_MWU for iological Process (BP) GO terms
commandArgs <- function(...) c("P_Recovery.vs.Ctrl.logP.txt", "BP")
source("GO_MWU.R")

# Run GO_MWU for Molecular Function GO terms
commandArgs <- function(...) c("P_Recovery.vs.Ctrl.logP.txt", "MF")
source("GO_MWU.R")

# Run GO_MWU for Cellular Component (CC) GO terms
commandArgs <- function(...) c("P_Recovery.vs.Ctrl.logP.txt", "CC")
source("GO_MWU.R")

setwd("analyses/GO_MWU/")
# Get list of all significant GO terms for contrast
P_HypRecovery.Ctrl.GO <- as.list(
  c(BP = "MWU_BP_P_Recovery.vs.Ctrl.logP.txt",
    MF = "MWU_MF_P_Recovery.vs.Ctrl.logP.txt",
    CC = "MWU_CC_P_Recovery.vs.Ctrl.logP.txt")
  ) %>% map(read.table, header = T) %>%
  bind_rows(.id = "ontology") %>%
  as_tibble() %>%
  filter(p.adj < 0.01) %>%
  select(ontology, nseqs, name, delta.rank, p.adj)

# Get list of individual genes with GO annotations and p-values for contrast
allgenepvals <- as.list(
  c(BP = "BP_P_Recovery.vs.Ctrl.logP.txt",
    MF = "MF_P_Recovery.vs.Ctrl.logP.txt",
    CC = "CC_P_Recovery.vs.Ctrl.logP.txt")
  ) %>% map(read_tsv) %>%
  bind_rows(.id = "ontology")

# For all significant GO terms, count how many genes with that GO term have raw pvalue < 0.05 (= "nsig")
P_HypRecovery.Ctrl.GO <- P_HypRecovery.Ctrl.GO %>%
  mutate(nsig = map_dbl(name, function(.) {
    filter(allgenepvals, name == ., -abs(value) <= log10(0.05)) %>%
    nrow(.)
  }))
  
# Print table
P_HypRecovery.Ctrl.GO %>%
  unite("genes", nsig, nseqs, sep = "/") %>%
  arrange(ontology, delta.rank) %>%
  knitr::kable()
```
# Check diff expression levels of hypoxia known genes for Ctrl vs hypoxia contrasts
## For abdominal 
```{r Hypoxia_genes A}
genesincommon_A <- DE_A %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(batch == "all", contrast %in% c("Hyp_T2h.Control","Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))
genesincommon_A_annot <- merge(genesincommon_A, annot_full[,c(1,5)], by="gene") %>% 
  distinct()

HipGenes <- c("HIF1A","PHD1","PHD2","PHD3","HIF1N","HRE","ZC3HF","HXK1","VGFR1","VGFR3","ARNT","EPAS1")

HIF1_A <- genesincommon_A_annot[genesincommon_A_annot$Hs_Gene %in% HipGenes, ]
gene.labs <- HIF1_A$Hs_Gene
names(gene.labs) <- HIF1_A$gene
HIF1_A <- HIF1_A[,-6]

# Get normalized counts for the gene
HIF1_counts <- data.frame(assay(vsd_A[which(rownames(assay(vsd_A)) %in% HIF1_A$gene), ]))
HIF1_counts <- rownames_to_column(HIF1_counts, var = "gene") %>%
  pivot_longer(-gene, names_to = "sample_ID", values_to = "count") %>%
  left_join(select(sdat0, sample_ID, treatment)) %>%
  mutate(gene = factor(gene, levels = levels(HIF1_A$gene))) %>%
  filter(treatment != "Control") %>%
  filter(treatment != "Reox") %>%
  filter(treatment != "Recovery")


# Get group log2foldchanges for plotting
gl2fc <- HIF1_A %>%
  pivot_longer(-gene, names_to = "treatment", values_to = "l2fc") %>%
  mutate(treatment = gsub(".Control", "", treatment))
  
# Get group means for plotting
HIF1_countsmean <- HIF1_counts %>%
  group_by(gene, treatment) %>%
  summarise(count = mean(count))

# Plot
plotgcounts <- HIF1_counts %>%
  ggplot(aes(x = treatment, y = pmax(0,log(count)))) +
  geom_jitter(width = 0.2, aes(color = treatment)) +
  geom_line(data = HIF1_countsmean, aes(group = gene)) +
  geom_text(data = gl2fc, aes(y = 2.4, label = round(l2fc, 2))) +
  facet_wrap(~ gene, labeller = labeller(gene=gene.labs)) +
  theme_custom() +
  theme(legend.position = "none") +
  labs(x = "", y = "log(Variance stabilized counts)") +
  scale_y_continuous(limits = c(0, 2.6))
plotgcounts


```

```{r Hypoxia_genes Pp}
genesincommon_P <- DE_Pp %>%
  unite("contrast", num, den, sep = ".") %>%
  filter(batch == "all", contrast %in% c("Hyp_T2h.Control","Hyp_T6h.Control", "LtH_6.Control", "LtH_7.Control")) %>%
  unnest(sig) %>%
  select(contrast, gene, log2FoldChange) %>%
  pivot_wider(names_from = contrast, values_from = log2FoldChange) %>%
  drop_na() %>%
  mutate(gene = factor(gene, levels = gene))
genesincommon_P_annot <- merge(genesincommon_P, annot_full[,c(1,5)], by="gene") %>% 
  distinct()

HipGenes <- c("HIF1A","PHD1","PHD2","PHD3","HIF1N","HRE","ZC3HF","HXK1","VGFR1","VGFR3","ARNT","EPAS1")

HIF1_P <- genesincommon_P_annot[genesincommon_P_annot$Hs_Gene %in% HipGenes, ]
## None of the target genes are in the Pp set

annot_full[annot_full$Hs_Gene %in% HipGenes,]
```

# KOG functional analysis 
## Batch 60 all tissues

```{r}
gene2kog <- gene2kog %>%
  filter(!KOG_class == "")

# Run KOG.MWU analysis on all contrasts
KOG <- DE %>%
  mutate(KOG = map(logP, ~ kog.mwu(., gene2kog)),
         # If a KOG term has <10 genes associated with it, set delta rank to zero
         KOG = map(KOG, ~ mutate(., delta.rank = ifelse(nseqs < 10, 0, delta.rank))))

# Generate all KOG delta rank tables
kogtables <- KOG %>%
  mutate(tissue= case_when(tissue == "A" ~ "abdominal", 
                           tissue == "Pp" ~ "Pleural/Pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(tissue = factor(tissue, levels = c("abdominal", "Pleural/Pedal", "all ganglia"))) %>%
  unnest(KOG) %>%
  select(tissue, num, den, term, delta.rank) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., tissue, delta.rank)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term"))

# Generate KOG p-value stars
kogpvals <- KOG %>%
  mutate(tissue= case_when(tissue == "A" ~ "abdominal", 
                           tissue == "Pp" ~ "Pleural/Pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(tissue = factor(tissue, levels = c("abdominal", "Pleural/Pedal", "all ganglia"))) %>%
  unnest(KOG) %>%
  select(tissue, num, den, term, padj) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., tissue, padj)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))

```

## KOG results for each tissue (batch 60)
```{r}
# Function to plot KOG delta ranks heatmap
KOGheatmap <- function(deltaranks, pvals, ...) {
  deltaranks <- as.matrix(deltaranks)
  paletteLength <- 100
  myColor <- colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(paletteLength)
  myBreaks <- c(seq(min(deltaranks), 0, length.out = ceiling(paletteLength/2) + 1), 
                seq(max(deltaranks)/paletteLength, max(deltaranks), 
                    length.out = floor(paletteLength/2)))
  pheatmap(mat = deltaranks, display_numbers = pvals, fontsize_number = 8,
           cluster_cols = FALSE, cluster_rows = FALSE,
           treeheight_row = 15, treeheight_col = 15,
           border_color = "white", scale = "none",
           color = myColor, breaks = myBreaks, ...)
}

KOG_T2vsC <- KOGheatmap(kogtables$Hyp_T2h.Control, kogpvals$Hyp_T2h.Control, main = "                     Hypoxia T2h vs. Control", 
                      fontsize = 8)
KOG_T6vsC <- KOGheatmap(kogtables$Hyp_T6h.Control, kogpvals$Hyp_T6h.Control, main = "                     Hypoxia T6h vs. Control", 
                      fontsize = 8)
KOG_T6dvsC <- KOGheatmap(kogtables$LtH_6.Control, kogpvals$LtH_6.Control, main = "                     Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_T7dvsC <- KOGheatmap(kogtables$LtH_7.Control, kogpvals$LtH_7.Control, main = "                     Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_RecoveryvsC <- KOGheatmap(kogtables$Recovery.Control, kogpvals$Recovery.Control, main = "                     Recovery vs. Control", 
                      fontsize = 8)
```

## Batch 71 all tissues

```{r KOG 71}

# Run KOG.MWU analysis on all contrasts
KOG <- DE_71 %>%
  mutate(KOG = map(logP, ~ kog.mwu(., gene2kog)),
         # If a KOG term has <10 genes associated with it, set delta rank to zero
         KOG = map(KOG, ~ mutate(., delta.rank = ifelse(nseqs < 10, 0, delta.rank))))

# Generate all KOG delta rank tables
kogtables <- KOG %>%
  mutate(tissue= case_when(tissue == "A" ~ "abdominal", 
                           tissue == "Pp" ~ "Pleural/Pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(tissue = factor(tissue, levels = c("abdominal", "Pleural/Pedal", "all ganglia"))) %>%
  unnest(KOG) %>%
  select(tissue, num, den, term, delta.rank) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., tissue, delta.rank)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term"))

# Generate KOG p-value stars
kogpvals <- KOG %>%
  mutate(tissue= case_when(tissue == "A" ~ "abdominal", 
                           tissue == "Pp" ~ "Pleural/Pedal", 
                           tissue == "all" ~ "all ganglia")) %>%
  mutate(tissue = factor(tissue, levels = c("abdominal", "Pleural/Pedal", "all ganglia"))) %>%
  unnest(KOG) %>%
  select(tissue, num, den, term, padj) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., tissue, padj)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))

```

## KOG results for each tissue (batch 71)
```{r}
# Function to plot KOG delta ranks heatmap
KOGheatmap <- function(deltaranks, pvals, ...) {
  deltaranks <- as.matrix(deltaranks)
  paletteLength <- 100
  myColor <- colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(paletteLength)
  myBreaks <- c(seq(min(deltaranks), 0, length.out = ceiling(paletteLength/2) + 1), 
                seq(max(deltaranks)/paletteLength, max(deltaranks), 
                    length.out = floor(paletteLength/2)))
  pheatmap(mat = deltaranks, display_numbers = pvals, fontsize_number = 8,
           cluster_cols = FALSE, cluster_rows = FALSE,
           treeheight_row = 15, treeheight_col = 15,
           border_color = "white", scale = "none",
           color = myColor, breaks = myBreaks, ...)
}

KOG_T2vsC <- KOGheatmap(kogtables$Hyp_T2h.Control, kogpvals$Hyp_T2h.Control, main = "                     Hypoxia T2h vs. Control", 
                      fontsize = 8)
KOG_T6vsC <- KOGheatmap(kogtables$Hyp_T6h.Control, kogpvals$Hyp_T6h.Control, main = "                     Hypoxia T6h vs. Control", 
                      fontsize = 8)
KOG_T6dvsC <- KOGheatmap(kogtables$LtH_6.Control, kogpvals$LtH_6.Control, main = "                     Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_T7dvsC <- KOGheatmap(kogtables$LtH_7.Control, kogpvals$LtH_7.Control, main = "                     Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_RecoveryvsC <- KOGheatmap(kogtables$Recovery.Control, kogpvals$Recovery.Control, main = "                     Recovery vs. Control", 
                      fontsize = 8)
```
## Abdominal both batches

```{r KOG abdominal}

# Run KOG.MWU analysis on all contrasts
KOG <- DE_A %>%
  mutate(KOG = map(logP, ~ kog.mwu(., gene2kog)),
         # If a KOG term has <10 genes associated with it, set delta rank to zero
         KOG = map(KOG, ~ mutate(., delta.rank = ifelse(nseqs < 10, 0, delta.rank))))

# Generate all KOG delta rank tables
kogtables <- KOG %>%
  mutate(batch= case_when(batch == 60 ~ "captive parents", 
                           batch == 71 ~ "wild parents", 
                           batch == "all" ~ "all")) %>%
  mutate(batch = factor(batch, levels = c("captive parents", "wild parents", "all"))) %>%
  unnest(KOG) %>%
  select(batch, num, den, term, delta.rank) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., batch, delta.rank)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term"))

# Generate KOG p-value stars
kogpvals <- KOG %>%
   mutate(batch= case_when(batch == 60 ~ "captive parents", 
                           batch == 71 ~ "wild parents", 
                           batch == "all" ~ "all")) %>%
  mutate(batch = factor(batch, levels = c("captive parents", "wild parents", "all"))) %>%
  unnest(KOG) %>%
  select(batch, num, den, term, padj) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., batch, padj)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))

```

## KOG results for each batch (abdominal)
```{r}
# Function to plot KOG delta ranks heatmap
KOGheatmap <- function(deltaranks, pvals, ...) {
  deltaranks <- as.matrix(deltaranks)
  paletteLength <- 100
  myColor <- colorRampPalette(rev(c("chocolate1","#FEE090","white", "cyan3","cyan")))(paletteLength)
  myBreaks <- c(seq(min(deltaranks), 0, length.out = ceiling(paletteLength/2) + 1), 
                seq(max(deltaranks)/paletteLength, max(deltaranks), 
                    length.out = floor(paletteLength/2)))
  pheatmap(mat = deltaranks, display_numbers = pvals, fontsize_number = 8,
           cluster_cols = FALSE, cluster_rows = FALSE,
           treeheight_row = 15, treeheight_col = 15,
           border_color = "white", scale = "none",
           color = myColor, breaks = myBreaks, ...)
}

KOG_T2vsC <- KOGheatmap(kogtables$Hyp_T2h.Control, kogpvals$Hyp_T2h.Control, main = "Hypoxia T2h vs. Control", 
                      fontsize = 8)
KOG_T6vsC <- KOGheatmap(kogtables$Hyp_T6h.Control, kogpvals$Hyp_T6h.Control, main = "Hypoxia T6h vs. Control", fontsize = 8, angle_col = 45)

KOG_T6dvsC <- KOGheatmap(kogtables$LtH_6.Control, kogpvals$LtH_6.Control, main = "Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_T7dvsC <- KOGheatmap(kogtables$LtH_7.Control, kogpvals$LtH_7.Control, main = "Hypoxia T7days vs. Control", 
                      fontsize = 8)
KOG_RecoveryvsC <- KOGheatmap(kogtables$Recovery.Control, kogpvals$Recovery.Control, main = "Recovery vs. Control", fontsize = 8, angle_col = 45)


```

## KOG results across all batches (abdominal)
```{r}
kogtable2 <- KOG %>%
  unnest(KOG) %>%
  filter(batch == "all") %>%
  select(num, den, term, delta.rank) %>%
  unite("contrast", num, den, sep = ".") %>%
  spread(contrast, delta.rank) %>%
  select(term, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control) %>%
  column_to_rownames("term")
  
kogpval2 <- KOG %>%
  unnest(KOG) %>%
  filter(batch == "all") %>%
  select(num, den, term, padj) %>%
  unite("contrast", num, den, sep = ".") %>%
  spread(contrast, padj) %>%
  select(term, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control) %>%
  column_to_rownames("term") %>%
  transmute_all(gtools::stars.pval)

KOG_abdominal <- KOGheatmap(
  select(kogtable2, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control),
  select(kogpval2, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control),
  labels_col = c(expression("2h_Hypoxia"),
                  expression("6h_Hypoxia"),
                  expression("6day_Hypoxia"),
                  expression("7day_Hypoxia"),
                  expression("Reoxigenation"),
                  expression("Recovery")),
  fontsize = 8, angle_col = 45)

ggsave(filename = "figures/KOG_abdominal.png", KOG_abdominal, width = 8, height = 4)
ggsave(filename = "figures/KOG_abdominal_Cvs6H.png", KOG_T6vsC, width = 6.5, height = 4)
ggsave(filename = "figures/KOG_abdominal_RecvsC.png", KOG_RecoveryvsC, width = 6.5, height = 4)
```

## Pleural/Pedal both batches

```{r KOG Pleural/Pedal}

# Run KOG.MWU analysis on all contrasts
KOG <- DE_Pp %>%
  mutate(KOG = map(logP, ~ kog.mwu(., gene2kog)),
         # If a KOG term has <10 genes associated with it, set delta rank to zero
         KOG = map(KOG, ~ mutate(., delta.rank = ifelse(nseqs < 10, 0, delta.rank))))

# Generate all KOG delta rank tables
kogtables <- KOG %>%
  mutate(batch= case_when(batch == 60 ~ "captive parents", 
                           batch == 71 ~ "wild parents", 
                           batch == "all" ~ "all")) %>%
  mutate(batch = factor(batch, levels = c("captive parents", "wild parents", "all"))) %>%
  unnest(KOG) %>%
  select(batch, num, den, term, delta.rank) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., batch, delta.rank)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term"))

# Generate KOG p-value stars
kogpvals <- KOG %>%
   mutate(batch= case_when(batch == 60 ~ "captive parents", 
                           batch == 71 ~ "wild parents", 
                           batch == "all" ~ "all")) %>%
  mutate(batch = factor(batch, levels = c("captive parents", "wild parents", "all"))) %>%
  unnest(KOG) %>%
  select(batch, num, den, term, padj) %>%
  split(list(.$num, .$den), drop = TRUE) %>%
  map(~ spread(., batch, padj)) %>%
  map(~ column_to_rownames(select(., -num, -den), "term")) %>%
  map(~ transmute_all(., gtools::stars.pval))

```

## KOG results for each batch (Pleural/Pedal)
```{r}
# Function to plot KOG delta ranks heatmap
KOGheatmap <- function(deltaranks, pvals, ...) {
  deltaranks <- as.matrix(deltaranks)
  paletteLength <- 100
  myColor <- colorRampPalette(rev(c("chocolate1","#FEE090","white", "cyan3","cyan")))(paletteLength)
  myBreaks <- c(seq(min(deltaranks), 0, length.out = ceiling(paletteLength/2) + 1), 
                seq(max(deltaranks)/paletteLength, max(deltaranks), 
                    length.out = floor(paletteLength/2)))
  pheatmap(mat = deltaranks, display_numbers = pvals, fontsize_number = 8,
           cluster_cols = FALSE, cluster_rows = FALSE,
           treeheight_row = 15, treeheight_col = 15,
           border_color = "white", scale = "none",
           color = myColor, breaks = myBreaks, ...)
}

KOG_T2vsC <- KOGheatmap(kogtables$Hyp_T2h.Control, kogpvals$Hyp_T2h.Control, main = "Hypoxia T2h vs. Control", 
                      fontsize = 8)
KOG_T6vsC <- KOGheatmap(kogtables$Hyp_T6h.Control, kogpvals$Hyp_T6h.Control, main = "Hypoxia T6h vs. Control", 
                      fontsize = 8, angle_col = 45)
KOG_T6dvsC <- KOGheatmap(kogtables$LtH_6.Control, kogpvals$LtH_6.Control, main = "Hypoxia T6days vs. Control", 
                      fontsize = 8)
KOG_T7dvsC <- KOGheatmap(kogtables$LtH_7.Control, kogpvals$LtH_7.Control, main = "Hypoxia T7days vs. Control", 
                      fontsize = 8)
KOG_RecoveryvsC <- KOGheatmap(kogtables$Recovery.Control, kogpvals$Recovery.Control, main = "Recovery vs. Control", 
                      fontsize = 8, angle_col = 45)


```

## KOG results across all batches (Pleural/Pedal)
```{r}
kogtable2 <- KOG %>%
  unnest(KOG) %>%
  filter(batch == "all") %>%
  select(num, den, term, delta.rank) %>%
  unite("contrast", num, den, sep = ".") %>%
  spread(contrast, delta.rank) %>%
  select(term, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control) %>%
  column_to_rownames("term")
  
kogpval2 <- KOG %>%
  unnest(KOG) %>%
  filter(batch == "all") %>%
  select(num, den, term, padj) %>%
  unite("contrast", num, den, sep = ".") %>%
  spread(contrast, padj) %>%
  select(term, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control) %>%
  column_to_rownames("term") %>%
  transmute_all(gtools::stars.pval)

KOG_Pp <- KOGheatmap(
  select(kogtable2, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control),
  select(kogpval2, Hyp_T2h.Control, Hyp_T6h.Control, LtH_6.Control, LtH_7.Control, Reox.Control, Recovery.Control),
  labels_col = c(expression("2h_Hypoxia"),
                  expression("6h_Hypoxia"),
                  expression("6day_Hypoxia"),
                  expression("7day_Hypoxia"),
                  expression("Reoxygenation"),
                  expression("Recovery")),
  fontsize = 8, angle_col = 45)

ggsave(filename = "figures/KOG_PleuralPedal.png", KOG_Pp, width = 8, height = 3.5)
ggsave(filename = "figures/KOG_PleuralPedal_Cvs6H.png", KOG_T6vsC, width = 6.5, height = 4)
ggsave(filename = "figures/KOG_PleuralPedal_RecvsC.png", KOG_RecoveryvsC, width = 6.5, height = 4)
```