---
title: "DMR stats and annotation"
author: "Javier Rodriguez Casariego"
date: "2023-06-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/JARC_2T/NSF-CREST_Postdoc/LongTerm_Hypoxia_exp/AplCal_hypoxia/WGBS/')
```

load libraries
```{r}
library(data.table)
library(gplots)
library(ggplot2) #install.packages('ggplot2')
library(dplyr)
library(broom)
library(RColorBrewer)
library(egg) #install.packages('egg')
library(purrr)
library(nlme)
library(vegan)
```

```{r set functions}
## ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10) %+replace%    #, base_family = "Arial"
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA),
      axis.text.x = element_text(angle=45, hjust=1, vjust = 1)#,
      #legend.title = element_text(size = 8), 
      #legend.text = element_text(size = 7)
    )
}

```
# Part 1:  Filter methylated regions for regions that have coverage in 75% of samples per group

read in data
```{r load data}

load("output/DMR_data.RData")

```

```{r prepare data for analysis}

# Build dataset for general stats 

perc_meth_by_sample_100 <- tidyr::gather(all_DMR_100bp[,7:ncol(all_DMR_100bp)], "Sample.ID", "perc.meth",1:24) %>%
  tidyr::spread(ID, perc.meth)
perc_meth_by_sample_100$Sample.ID <- gsub("methylation_level_","", perc_meth_by_sample_100$Sample.ID)

colnames(meta) <- c("Sample.ID", "Animal_ID", "Tissue", "Treatment", "Batch")
perc_meth_by_sample_100 <- merge(meta[,c(1,4:5)], perc_meth_by_sample_100, by = "Sample.ID")

```
## Run permanova to evaluate methylation interaction with site and genotype

```{r}
set.seed(1234)
adonis_treat <- adonis2(perc_meth_by_sample_100[, c(4:ncol(perc_meth_by_sample_100))] ~ Treatment, strata = perc_meth_by_sample_100$Batch, data = perc_meth_by_sample_100, permutations = 9999, method = "binomial", na.rm = TRUE)

adonis_treatBatch <- adonis2(perc_meth_by_sample_100[, c(4:ncol(perc_meth_by_sample_100))] ~ Treatment*Batch, data = perc_meth_by_sample_100, permutations = 9999, method = "manhattan",  na.rm = TRUE)

adonis_treat
adonis_treatBatch

```


# Part 2: Run group statistics on regions to find regions that are significantly different among groups

reformat data for calculating group effect
```{r}
#reformat all for t test
all_DMR_100_STACKED <- tidyr::gather(all_DMR_100bp[,7:ncol(all_DMR_100bp)], "Sample.ID", "perc.meth",1:24)

#simplify sample name
all_DMR_100_STACKED$Sample.ID <- gsub("methylation_level_","", all_DMR_100_STACKED$Sample.ID)

#merge with meta data
all_DMR_100_STACKED <- merge(all_DMR_100_STACKED, meta, by = "Sample.ID")

naive_DMR_100_STACKED <- all_DMR_100_STACKED[which(all_DMR_100_STACKED$Batch == "60"),]
preexposed_DMR_100_STACKED <- all_DMR_100_STACKED[which(all_DMR_100_STACKED$Batch == "71"),]
```

plot % meth dist.
```{r}
a <- ggplot(all_DMR_100_STACKED) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

a

a2 <- ggplot(all_DMR_100_STACKED) + 
  geom_histogram(aes(perc.meth, group = Batch, color = Batch,fill = Batch), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

a2

a3 <- ggplot(naive_DMR_100_STACKED) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

a3

a4 <- ggplot(preexposed_DMR_100_STACKED) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

a4
```

arc sin sqrt transform the data
```{r}
#arcsin sqrt transformation function
asinTransform <- function(p) { asin(sqrt(p))}

all_DMR_100_STACKED_asin <- all_DMR_100_STACKED
all_DMR_100_STACKED_asin$perc.meth <- asinTransform(all_DMR_100_STACKED_asin$perc.meth)
naive_DMR_100_STACKED_asin <- all_DMR_100_STACKED_asin[which(all_DMR_100_STACKED_asin$Batch == "60"),]
preexposed_DMR_100_STACKED_asin <- all_DMR_100_STACKED_asin[which(all_DMR_100_STACKED_asin$Batch == "71"),]

```

plot transformed dist.
```{r}
b <- ggplot(all_DMR_100_STACKED_asin) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

b

b2 <- ggplot(all_DMR_100_STACKED_asin) + 
  geom_histogram(aes(perc.meth, group = Batch, color = Batch,fill = Batch), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

b2

b3 <- ggplot(naive_DMR_100_STACKED_asin) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

b3

b4 <- ggplot(preexposed_DMR_100_STACKED_asin) + 
  geom_histogram(aes(perc.meth, group = Treatment, color = Treatment,fill = Treatment), bins = 20, position = "identity", alpha = 0.5) +   theme_bw() + 
  xlab("fraction methylated CpGs") + 
  theme(text = element_text(size=8),plot.title = element_text(size = 8,face = "bold"),legend.title = element_text(size = 4),legend.text = element_text(size = 4))

b4

jpeg("output/DMR_100_percMeth_histograms.jpeg", width = 8, height = 16, units = "in", res = 300)
ggarrange(a, b, a2, b2, a3, b3, a4, b4, nrow = 4)
dev.off()

```
## Run anova on TRANSFORMED data to assess group differences for each DMR

## One-way anova
**Hypothesis: there is no effect of batch 
```{r}
all_DMR_100_1way_aov <- all_DMR_100_STACKED_asin %>% group_by(ID) %>%
do(meth_aov_models = aov(perc.meth~Batch, data =  . ))
#summarize ANOVA data
all_DMR_100_1way_aov_modelsumm  <- all_DMR_100_1way_aov %>% ungroup %>% 
    pull(meth_aov_models) %>% 
    map_dfr(tidy, .id = 'grp')
#spread out pvalues
all_DMR_100_1way_aov_modelsumm_wide <- data.frame(tidyr::pivot_wider(all_DMR_100_1way_aov_modelsumm, names_from = term, values_from = c("df", "sumsq", "meansq","statistic","p.value"),names_sep = "_" ))
all_DMR_100_1way_aov_modelsumm_wide <- cbind(all_DMR_100_1way_aov[,1], all_DMR_100_1way_aov_modelsumm_wide[,-c(1)])
length(all_DMR_100_1way_aov_modelsumm_wide[which(all_DMR_100_1way_aov_modelsumm_wide$p.value_Batch <= 0.05), 1])

```

## Two-way anova
**Hypotehsis: there is no effect of batch or treatment or its interaction
```{r}
all_DMR_100_2way_aov <- all_DMR_100_STACKED_asin %>% group_by(ID) %>%
do(meth_aov_models = aov(perc.meth ~Batch*Treatment, data =  . ))
#summarize ANOVA data
all_DMR_100_2way_aov_modelsumm  <- all_DMR_100_2way_aov %>% ungroup %>% 
    pull(meth_aov_models) %>% 
    map_dfr(tidy, .id = 'grp')

#spread out pvalues
all_DMR_100_2way_aov_modelsumm_wide <- data.frame(tidyr::pivot_wider(all_DMR_100_2way_aov_modelsumm, names_from = term, values_from = c("df", "sumsq", "meansq","statistic","p.value"),names_sep = "_" ))
all_DMR_100_2way_aov_modelsumm_wide <- cbind(all_DMR_100_2way_aov[,1], all_DMR_100_2way_aov_modelsumm_wide[,-c(1)])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Treatment <= 0.05), 1])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch <= 0.05), 1])
length(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch.Treatment <= 0.05), 1])
```

#write out DMR bed file for 1way anova

```{r}
all_DMR_100_1way_aov_modelsumm_wide$scaffold <- gsub("\\:.*","", all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide$start <- gsub(".*\\:","", all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide$start <- gsub("-.*","", all_DMR_100_1way_aov_modelsumm_wide$start)
all_DMR_100_1way_aov_modelsumm_wide$end <- gsub(".*-","",  all_DMR_100_1way_aov_modelsumm_wide$ID)
all_DMR_100_1way_aov_modelsumm_wide <- all_DMR_100_1way_aov_modelsumm_wide[,c(1,(ncol(all_DMR_100_1way_aov_modelsumm_wide)-2):ncol(all_DMR_100_1way_aov_modelsumm_wide),2:(ncol(all_DMR_100_1way_aov_modelsumm_wide)-3))]
#order by pvalue
all_DMR_100_1way_aov_modelsumm_wide <- all_DMR_100_1way_aov_modelsumm_wide[order(all_DMR_100_1way_aov_modelsumm_wide$p.value_Batch),]
write.csv(all_DMR_100_1way_aov_modelsumm_wide,"output/aov_batch.csv", row.names = FALSE, quote = FALSE)

```

### Export file for feature analysis

generate bedfiles to be used in bedtools intersect to determine features that overlap with regions
```{r}
#subset model data for scaffold start and end positions for DMRs significant at p < 0.05
DMR_100_sig.bed <- all_DMR_100_1way_aov_modelsumm_wide[which(all_DMR_100_1way_aov_modelsumm_wide$p.value_Batch < 0.05), c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig.bed$test <- "batch"

#write out data
write.table(DMR_100_sig.bed, "output/aov_0.05_batch.bed", sep = "\t",row.names = F, col.names = F, quote = F)

```

#write out DMR bed file for 2way anova

```{r}
all_DMR_100_2way_aov_modelsumm_wide$scaffold <- gsub("\\:.*","", all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide$start <- gsub(".*\\:","", all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide$start <- gsub("-.*","", all_DMR_100_2way_aov_modelsumm_wide$start)
all_DMR_100_2way_aov_modelsumm_wide$end <- gsub(".*-","",  all_DMR_100_2way_aov_modelsumm_wide$ID)
all_DMR_100_2way_aov_modelsumm_wide <- all_DMR_100_2way_aov_modelsumm_wide[,c(1,(ncol(all_DMR_100_2way_aov_modelsumm_wide)-2):ncol(all_DMR_100_2way_aov_modelsumm_wide),2:(ncol(all_DMR_100_2way_aov_modelsumm_wide)-3))]
#order by pvalue
all_DMR_100_2way_aov_modelsumm_wide <- all_DMR_100_2way_aov_modelsumm_wide[order(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch.Treatment),]
write.csv(all_DMR_100_2way_aov_modelsumm_wide,"output/aov2way_all.csv", row.names = FALSE, quote = FALSE)
```

### Export file for feature analysis

generate bedfiles to be used in bedtools intersect to determine features that overlap with regions
```{r}
#subset model data for scaffold start and end positions for DMRs significant at p < 0.05
DMR_100_sig_treatment_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Treatment < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_treatment_2way.bed$test <- "Treatment"

#write out data
write.table(DMR_100_sig_treatment_2way.bed, "output/aov2way_0.05_treatment.bed", sep = "\t",row.names = F, col.names = F, quote = F)

DMR_100_sig_batch_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_batch_2way.bed$test <- "Batch"

#write out data
write.table(DMR_100_sig_batch_2way.bed, "output/aov2way_0.05_batch.bed", sep = "\t",row.names = F, col.names = F, quote = F)

DMR_100_sig_interaction_2way.bed <- all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch.Treatment < 0.05),c("scaffold","start", "end")]

#add column denoting tested effect
DMR_100_sig_interaction_2way.bed$test <- "batch:treatment"

#write out data
write.table(DMR_100_sig_interaction_2way.bed, "output/aov2way_0.05_batch:treatment.bed", sep = "\t",row.names = F, col.names = F, quote = F)
```

plot heatmap of Anova for batch:treatment p.val > 0.05 sig DMRs
```{r}
#create matrix for one way ANOVA results
aov_0.05_interaction_DMR_m <- as.matrix(all_DMR_100bp[,7:30])
rownames(aov_0.05_interaction_DMR_m) <- all_DMR_100bp$ID
aov_0.05_interaction_DMR_m <- aov_0.05_interaction_DMR_m[which(rownames(aov_0.05_interaction_DMR_m) %in% pull(all_DMR_100_2way_aov_modelsumm_wide[which(all_DMR_100_2way_aov_modelsumm_wide$p.value_Batch.Treatment<= 0.05),],ID)),]

#remove extra text from column names
colnames(aov_0.05_interaction_DMR_m) <- gsub("methylation_level_","",colnames(aov_0.05_interaction_DMR_m))

###Visualize group means fo all
#calculate group means
MeanNaiveCtrol <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 60 & meta$Treatment=="Control"), 1]], na.rm = TRUE)
MeanNaive6h <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 60 & meta$Treatment=="H_6h"), 1]], na.rm = TRUE)
MeanNaive6d <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 60 & meta$Treatment=="H_6-days"), 1]], na.rm = TRUE)
MeanNaiveRecovery <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 60 & meta$Treatment=="Recovery"), 1]], na.rm = TRUE)

MeanPreCtrol <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 71 & meta$Treatment=="Control"), 1]], na.rm = TRUE)
MeanPre6h <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 71 & meta$Treatment=="H_6h"), 1]], na.rm = TRUE)
MeanPre6d <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 71 & meta$Treatment=="H_6-days"), 1]], na.rm = TRUE)
MeanPreRecovery <- rowMeans(aov_0.05_interaction_DMR_m[,colnames(aov_0.05_interaction_DMR_m) %in% meta[which(meta$Batch == 71 & meta$Treatment=="Recovery"), 1]], na.rm = TRUE)

#bind all group means together
aov_0.05_interaction_DMR_mean <- as.matrix(data.frame(cbind(MeanNaiveCtrol, MeanNaive6h, MeanNaive6d, MeanNaiveRecovery, MeanPreCtrol, MeanPre6h, MeanPre6d, MeanPreRecovery)))

#plot site mean comparison
colnames(aov_0.05_interaction_DMR_mean) <- c("naive_Ctrol", "naive_H_6h","naive_H_6d","naive_Recovery",
                                       "pre-exposed_Ctrol","pre-exposed_H_6h","pre-exposed_H_6d","pre-exposed_Recovery")

jpeg("output/aov_0.05_100_DMR_heatmap.jpg", width = 1000, height = 1000)
heatmap.2(aov_0.05_interaction_DMR_mean, margins = c(10,10), cexCol = , distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=NA, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(aov_0.05_interaction_DMR_mean),rowsep=1:nrow(aov_0.05_interaction_DMR_mean),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()


genet_color <- data.frame(sample=colnames(aov_0.05_site_DMR_m), genet=meta[meta$Sample.ID %in% genet_color$sample, 5], color = color <-c("#F8766D", "#F8766D", "#FF64B0", "#FF64B0", "#00B4F0", "#00B4F0", "#00C08B", "#00C08B", "#619CFF", "#619CFF", "#F564E3", "#F564E3", "#DE8C00", "#DE8C00", "#C77CFF", "#C77CFF", "#00BA38", "#00BA38", "#DE8C00", "#DE8C00", "#00BFC4", "#00BFC4", "#B79F00", "#B79F00", "#7CAE00", "#7CAE00"))


# set the custom distance and clustering functions
hclustfunc <- function(x) hclust(x,method = 'average')
distfunc <- function(x) as.dist(1 - cor(t(x), use = "pa"))
key.xtickfun <- function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)
  }

jpeg("DMR.aov_0.05_100_bysample_heatmap.jpg", width = 1000, height = 1000)
heatmap.2(aov_0.05_site_DMR_m, margins = c(10,10), hclustfun=hclustfunc, distfun=distfunc, key.xtickfun = key.xtickfun, cexCol = 2, Colv=T, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(aov_0.05_site_DMR_m),rowsep=1:nrow(aov_0.05_site_DMR_m),keysize=0.5, key.par = list(cex=1), ColSideColors=genet_color$color)
dev.off()


```



##plot heatmap of sig DMRs AOV_1way overlaping with gene regions
```{r}
#create matrix for one way ANOVA results
gene_DMR_m <- as.matrix(all_DMR_100bp[,7:32])
rownames(gene_DMR_m) <- all_DMR_100bp$ID

gene_overlap <- read.csv("analyses/Genomic_features/DMR_genomic_features/aov_0.05_site_gene_trans.txt", sep = "\t", header = FALSE)
gene_overlap$ID <- paste(gene_overlap$V1,":", gene_overlap$V2,"-",gene_overlap$V3, sep = "")
gene_overlap$ID <- gsub("__.*__.*:",":",gene_overlap$ID)

gene_DMR_m <- gene_DMR_m[which(rownames(gene_DMR_m) %in% gene_overlap$ID),]

#remove extra text from column names
colnames(gene_DMR_m) <- gsub("methylation_level_","",colnames(gene_DMR_m))

###Visualize group means fo all
#calculate group means
MeanDeep <- rowMeans(gene_DMR_m[,grep("_D", colnames(gene_DMR_m))], na.rm = TRUE)
MeanShallow <- rowMeans(gene_DMR_m[,grep("_S", colnames(gene_DMR_m))], na.rm = TRUE)

#bind all group means together
gene_DMR_mean <- as.matrix(data.frame(cbind(MeanDeep,MeanShallow)))

#plot site mean comparison
colnames(gene_DMR_mean) <- c("Deep", "Shallow")
jpeg("output/gene.aov_0.05_100_DMR_heatmap.jpg", width = 600, height = 1000)
heatmap.2(gene_DMR_mean,margins = c(10,20), cexCol = 2, distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=NA, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(gene_DMR_mean),rowsep=1:nrow(gene_DMR_mean),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()

#plot each individual sample
jpeg("output/DMR.gene_aov_0.05_100_bysample_heatmap.jpg", width = 1000, height = 1000)
heatmap.2(gene_DMR_m,margins = c(10,10), cexCol = 2, distfun = function(x) as.dist(1 - cor(t(x), use = "pa")), hclustfun = function(x) hclust(x,method = 'average'),key.xtickfun = function() {
  breaks = pretty(parent.frame()$breaks)
  #breaks = breaks[c(1,length(breaks))]
  list(at = parent.frame()$scale01(breaks),labels = breaks)},Colv=TRUE, col= rev(colorRampPalette(brewer.pal(10, "RdYlBu"))(256)), density.info = "none", trace = "none", scale = "row", labRow = FALSE,sepwidth=c(0.01,0.01),sepcolor="white",colsep=1:ncol(gene_DMR_m),rowsep=1:nrow(gene_DMR_m),lmat = rbind(c(0,3),c(2,1),c(4,0)),keysize=0.5, key.par = list(cex=1),lhei=c(1.5,4,1), lwid = c(1.5,4))
dev.off()

```