---
title: "smalRNA_annotation"
author: Javier A. Rodriguez Casariego
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
---

# Setup

```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/JARC_2T/NSF-CREST_Postdoc/LongTerm_Hypoxia_exp/AplCal_hypoxia/miRNA/')
```


```{bash process alignments}
## Map sRNA-seq data to the genome with bowtie

trimmed_files="/scratch/jeirinlo/jrodr979/sncRNA_ACAL/Trimmed_files"
genome="/scratch/jeirinlo/jrodr979/ACAL_meth/genome/unplaced.scaf.fna"
output="/scratch/jeirinlo/jrodr979/sncRNA_ACAL/20240404_Bowtie_mapping"
# Build bowtie index

#bowtie-build ${genome} ACAL_index

# Run alignments

for f in ${trimmed_files}/*.fastq 
do
STEM=$(basename "${f}" -trimmed.fastq)
bowtie \
-v 2 \
--best \
-m 50 \
-p $SLURM_CPUS_ON_NODE \
-S ACAL_index \
"${f}" > "${STEM}".sam
done

## Convert sam files to bam
for f in *.sam
do
samtools view -bS ${f} > ${f/.sam}.bam
done

## get bed file of all mapped reads
for f in *.bam
do
bedtools bamtobed -i ${f} > ${f/.bam}.bed
done

#Obtain sequence lengths for each chromosome
awk '$0 ~ ">" {print c; c=0;printf substr($0,2,100) "\t"; } $0 !~ ">" {c+=length($0);} END { print c; }' \
/scratch/jeirinlo/jrodr979/ACAL_meth/genome/unplaced.scaf.fna |\
awk '{print $1"\t"$NF}' > ACAL.genome_assembly-sequence-lengths.txt

# Sort bed files before merging
for f in *.bed
do
sortBed -faidx ACAL.genome_assembly-sequence-lengths.txt -i ${f} > ${f/.bed}.sorted.bed
done

# merge all bed files to get track of all small RNA's across all samples. only merge locations that overlap at least in 10 bases. This will allow to merge tracks that deviate in one or two bases without merging neighboring ones. piRNAs for example can be organized in tandems. 

cat *.sorted.bed | sortBed -faidx ACAL.genome_assembly-sequence-lengths.txt -i | bedtools merge -s -d -10 > sRNA_merged.bed


```

```{r count read lenghts and separate potential miRNA, piRNA, etc}

library(gplots)
library(ggplot2)

merged.bed <- read.delim("data/sRNA_merged.bed", header = FALSE, sep = "\t")
colnames(merged.bed) <- c("chr", "start", "end")
merged.bed$length <- merged.bed$end - merged.bed$start
str(merged.bed)

small.bed <- merged.bed[which(merged.bed$length < 36),]
range(small.bed$length)
a <- ggplot(small.bed) + 
  geom_histogram(aes(length), binwidth = 1) + theme_bw()

a

# there is a clear bimodal distribution with two peaks, at 18-24 (miRNA), 25-35 (piRNA + other)
ggsave("output/smRNA_histogram_length.png", a, width = 4, height = 4)

```
Peak between 16-24 corresponds mostly to miRNA and peak between 25-35 is likely including the piRNA

```{bash ProTrac pipeline}
# This pipeline was described in https://www.nature.com/articles/s42003-018-0141-4#Sec9 

# Remove redundant reads from trimmed fastqs, keep reads between 25 and 35 nt, remove low complexity reads and change the format for SeqMap alignment

for f in ${trimmed}/*.fastq
do
perl TBr2_collapse.pl -i ${f} -o ${f/-trimmed.fastq}.collapsed
perl TBr2_length-filter.pl -i ${f/-trimmed.fastq}.collapsed -o ${f/-trimmed.fastq}.collapsed.filt -min 25 -max 35
perl TBr2_duster.pl -i ${f/-trimmed.fastq}.collapsed.filt
done

# concatenate reads in a single file

cat *.collapsed.filt.no-dust > filt.merged.fq

# Filter rRNA, tRNA, snRNAs, miRNAs, and snoRNAs reads (based in annotations on gff file non of these are piwi or related. 

# build a reference dataset with all annotated RNAs that want to exclude

bedtools getfasta \
-fi ${genome}/unplaced.scaf.fasta \
-bed ${genome}/genomic_features/Acal.GFFannotation.<snoRNA/rRNA/tRNA/snRNA>.gff \
-fo ${genome}/<snoRNA/rRNA/tRNA/snRNA>.fasta 

cat *RNA.fasta > ${genome}/otherSmallRNAs.fasta

## use sortmerna to clear reads that match the fasta
sortmerna \
--ref ${genome}/otherSmallRNAs.fasta \
--reads filt.merged.fq \
--workdir /scratch/jeirinlo/jrodr979/sncRNA_ACAL/20240409_piRNA/20240415_sortMerna/ \
--fastx \
--other

mv /scratch/jeirinlo/jrodr979/sncRNA_ACAL/20240409_piRNA/20240415_sortMerna/out/other.fq merged_preproc

## Align clean reads to the genome

# simplify fasta headers on genome to include only the scaffold or chr name. proTRAC dont go well with long headers

sed 's/^>\([^ ]*\).*/>\1/' unplaced.scaf.fna > unplaced.scaf.fasta 

# option 1

perl sRNAmapper.pl \
-input merged_preproc \
-genome ${genome}/unplaced.scaf.fasta \
-alignments best

# option 2

seqmap 3 seq.fasta.collapsed.no-dust genome.fas seq.fasta.collapsed.no-dust.map /output_all_matches

perl seqmap_filter.pl seq.fasta.collapsed.no-dust.map 1 2

# define origin for multiple mapping reads

perl reallocate.pl merged_preproc.map 10000 1000 b 0 

# run proTRAC
perl proTRAC_2.4.2.pl \
-map merged_preproc.map \
-genome ${genome}/unplaced.scaf.fna \
-repeatmasker ${genome}/unplaced.scaf.fna.out \
-geneset ${genome}/genomic.gtf 



```


